---
title: "Explore germline variant distributions"
output: html_notebook
author: Jo Lynne Rokita & Ryan Corbett
date: 2023
editor_options: 
  chunk_output_type: inline
---

# Load libraries
```{r load libraries, set up directories}
suppressPackageStartupMessages({
  library(ggplot2)
  library(tidyverse)
})

# Set up directories
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "variant-distribution")
input_dir <- file.path(analysis_dir, "input")
plot_dir <- file.path(analysis_dir, "plots/")
results_dir <- file.path(analysis_dir, "results")

# If the directory does not exist, create it
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

if (!dir.exists(plot_dir)) {
  dir.create(plot_dir, recursive = TRUE)
}
```

# Load files
```{r load figure theme, files}
# source publication theme
source(file.path(root_dir, "figures", "theme.R"))

# read in cpg and driver lists
cpg <- read_lines(file.path(root_dir, "analyses", "oncokb-annotation", "input", "cpg.txt"))

# somatic drivers by cancer group
embryonal_genes <- read_lines(file.path(input_dir, "embryonal-tumor_goi_list.tsv"))
hgg_genes <- read_lines(file.path(input_dir, "hgat_goi_list.tsv"))
lgg_genes <- read_lines(file.path(input_dir, "lgat_goi_list.tsv"))
other_genes <- read_lines(file.path(input_dir, "other_goi_list.tsv"))

# cat somatic lists
somatic_drivers <- unique(c(embryonal_genes, hgg_genes, lgg_genes, other_genes))

# read in histology file
hist <- read_tsv(file.path(root_dir, 
                           "analyses", 
                           "collapse-tumor-histologies", 
                           "results",
                           "germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv"), show_col_types = FALSE)
```


# Merge clin with germline plp file
```{r data wrangling}
plp <- read_tsv(file.path(data_dir, "pbta-merged-plp-variants-autogvp-abridged.tsv"), 
                show_col_types = FALSE) %>%
  filter(Kids_First_Biospecimen_ID_normal %in% hist$Kids_First_Biospecimen_ID_normal) %>%
  # unique one gene per patient if multiple variants (in this case we have 1 normal per pt, so can use bs_id)
  dplyr::select(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, variant_classification_vep) %>%
  distinct(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, .keep_all = T) %>%
  # Consolidate variant classifications, and pick most severe when multiple reported (missense > stop > frameshift > splice)
  dplyr::mutate(variant_classification_vep = case_when(
    grepl("missense_variant", variant_classification_vep) ~ "missense_variant",
    grepl("stop_gained", variant_classification_vep) ~ "stop_gained",
    grepl("frameshift_variant", variant_classification_vep) ~ "frameshift_variant",
    grepl("splice_donor", variant_classification_vep) ~ "splice_donor_variant",
    grepl("splice_acceptor", variant_classification_vep) ~ "splice_acceptor_variant",
    grepl("splice", variant_classification_vep) ~ "splice_region_variant",
    TRUE ~ variant_classification_vep
  ))

# how many CPG have P/LP?
plp_cpg <- plp %>%
  filter(gene_symbol_vep %in% cpg)

length(unique(plp_cpg$gene_symbol_vep))

# how many somatic drivers have P/LP?
plp_somatic <- plp %>%
  filter(gene_symbol_vep %in% somatic_drivers)

length(unique(plp_somatic$gene_symbol_vep))

# how many somatic drivers which are not CPG have P/LP?
plp_som_not_cpg <- plp %>%
  filter(gene_symbol_vep %in% plp_somatic$gene_symbol_vep & !gene_symbol_vep %in% plp_cpg$gene_symbol_vep) %>%
  left_join(hist) %>%
  write_tsv(file.path(results_dir, "plp-variants-in-somatic-drivers-not-cpg.tsv"))
  
length(unique(plp_som_not_cpg$gene_symbol_vep))
sort(unique(plp_som_not_cpg$gene_symbol_vep))

```

Create barplots of patient P/LP carrier count by CPG and histology
```{r}
plplists <-  list("cpg" = plp_cpg, "somatic_drivers" = plp_somatic, "somatic_not_cpg" = plp_som_not_cpg)               

for (each in names(plplists)){
  
  # add titles to print
  if (each == "cpg"){
    x_axis_title <- "Cancer predisposition gene (CPG)"
    fig_width <- 4200
  }
  if (each == "somatic_drivers"){
    x_axis_title <- "CPG or somatic driver gene"
    fig_width <- 4000
  }  
  if (each == "somatic_not_cpg"){
    x_axis_title <- "Somatic driver gene"
    fig_width <- 2000
  }
  
  plot_df <- plplists[[each]] %>%
    left_join(hist) 

  # get N per histology per gene
  n_per_hist <- plot_df %>%
    count(gene_symbol_vep, plot_group) %>%
    dplyr::rename(hist_n = n)

  # create N per gene desc order for plot
  n_per_gene <- plot_df %>%
    left_join(hist) %>%
    count(gene_symbol_vep) %>%
    dplyr::rename(gene_n = n) %>%
    arrange(desc(gene_n)) %>%
    # create plot order desc by gene
    mutate(plot_order = row_number()) %>%
    # add back histology info
    left_join(n_per_hist) %>%
    left_join(plot_df[,c("plot_group", "plot_group_hex")]) %>%
    distinct()

  # palette
  plot_group_palette <- n_per_gene$plot_group_hex
  names(plot_group_palette) <- n_per_gene$plot_group


dist_plot <- ggplot(data = n_per_gene, aes(x = reorder(gene_symbol_vep, plot_order), y = hist_n, fill = plot_group)) +
  geom_col(col = "black", size = 0.4) +
  scale_fill_manual(values = plot_group_palette, breaks = sort(names(plot_group_palette))) + 
  ylab("Number of patients with P/LP variants") +
  xlab(x_axis_title) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ylim(0,20) +
  labs(fill = "Tumor diagnosis")

print(dist_plot)

dev.set(dev.next())
tiff(file.path(paste0(plot_dir, each, "-variant-distribution-by-histology.tiff")), height = 1800, width = fig_width, res = 300)
print(dist_plot)
dev.off()

}

```

Create barplots of patient P/LP carrier count by CPG and variant classification
```{r}
plplists <-  list("cpg" = plp_cpg, "somatic_drivers" = plp_somatic, "somatic_not_cpg" = plp_som_not_cpg)


vc_colors <- tibble(variant_classification_vep = unique(plp_cpg$variant_classification_vep),
                    vc_hex = RColorBrewer::brewer.pal(n = 11, name = 'Paired'))

for (each in names(plplists)){
  
  # add titles to print
  if (each == "cpg"){
    x_axis_title <- "Cancer predisposition gene (CPG)"
    fig_width <- 4300
  }
  if (each == "somatic_drivers"){
    x_axis_title <- "CPG or somatic driver gene"
    fig_width <- 4000
  }  
  if (each == "somatic_not_cpg"){
    x_axis_title <- "Somatic driver gene"
    fig_width <- 2000
  }
  
  plot_df <- plplists[[each]] %>%
    left_join(hist) 

  # get N per variant classification per gene
  n_per_vc <- plot_df %>%
    count(variant_classification_vep, gene_symbol_vep) %>%
    dplyr::rename(vc_n = n)

  # create N per gene desc order for plot
  n_per_gene <- plot_df %>%
    count(gene_symbol_vep) %>%
    dplyr::rename(gene_n = n) %>%
    arrange(desc(gene_n)) %>%
    # create plot order desc by gene
    mutate(plot_order = row_number()) %>%
    # add back variant classification info
    left_join(n_per_vc) %>%
    left_join(vc_colors) %>%
    distinct()

  vc_palette <- n_per_gene$vc_hex
  names(vc_palette) <- n_per_gene$variant_classification_vep

  dist_plot <- ggplot(data = n_per_gene, aes(x = reorder(gene_symbol_vep, plot_order), y = vc_n, fill = variant_classification_vep)) +
    geom_col(col = "black", size = 0.4) +
    scale_fill_manual(values = vc_palette, breaks = sort(names(vc_palette))) + 
    ylab("Number of patients with P/LP variants") +
    xlab(x_axis_title) +
    theme_Publication() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    ylim(0,20) +
    labs(fill = "Variant Classification")

  print(dist_plot)
  
  dev.set(dev.next())
  tiff(file.path(paste0(plot_dir, each, "-variant-distribution-by-variant-classification.tiff")), height = 1800, width = fig_width, res = 300)
  print(dist_plot)
  dev.off()

}

```
Plot variant allele frequency using full AutoGVP output (contains allele depths)
```{r}
plp_full <- read_tsv(file.path(data_dir,
                               "pbta-merged-plp-variants-autogvp-full.tsv")) %>%
  dplyr::mutate(AF = AD_alt/(AD_alt + AD_ref)) %>%
  dplyr::filter(gene_symbol_vep %in% cpg)

plp_full %>%
  ggplot(aes(x = AF)) +
  geom_histogram(color = "black", fill = "cadetblue") +
  xlim(c(0,1.0)) + 
  xlab("Variant allele frequency") +
  theme_Publication()

ggsave(file.path(plot_dir, "cpg-variant-allele-freq-hist.tiff"),
       width = 4, height = 6)

```

# print session info
```{r session info}
sessionInfo()
```





