---
title: "Assess P/LP variant- and splicing-associated somatic gene expression changes in PBTA germline cohort"
output: 
  html_notebook:
  toc: TRUE
toc_float: TRUE
author: Ryan Corbett
---

This script assesses somatic gene expression in germline CPG P/LP carriers by:

1) Variant classification, with specific interest in splice variant-associated gene expression
2) Level of proximal alternative splicing (single exon skipping, intron retention)

# Load libraries

```{r load libraries, set up directories}
suppressPackageStartupMessages({
  library(ggplot2)
  library(tidyverse)
  library(vroom)
  library(circlize)
  library(ggpubr)
  library(ggrepel)
  library(ggsci)
})

# Set up directories
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "alternative-splicing")
input_dir <- file.path(analysis_dir, "input")
results_dir <- file.path(analysis_dir, "results")
plot_dir <- file.path(analysis_dir, "plots")

# If the directory does not exist, create it
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

# Call plotting theme
source(file.path(root_dir, "figures", "theme.R"))
source(file.path(analysis_dir, "util", "splicing_plot_functions.R"))

set.seed(2024)
```

Set input file paths

```{r}
plp_file <- file.path(data_dir, 
                      "pbta-merged-plp-variants-autogvp-abridged.tsv")

plp_lowVAF_file <- file.path(data_dir,
                             "pbta-merged-plp-variants-autogvp-abridged-lowVAF-cpgs.tsv")

plp_sv_file <- file.path(data_dir,
                         "pbta_germline_svs.tsv")

# cpg file
cpg_file <- file.path(root_dir, "analyses", "oncokb-annotation", 
                      "input", "cpg.txt")

expr_file <- file.path(root_dir, "analyses",
                       "two-hits", "results", "pbta-germline-gene-expr-zscores.tsv")

prot_file <- file.path(root_dir, "analyses",
                       "two-hits", "results",
                       "cptac-hope-cpg-proteomics-zscores.tsv")

hist_cohort_file <- file.path(root_dir, "analyses", 
                              "collapse-tumor-histologies", "results", 
                              "germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv")
```

# Wrangle data

```{r}
cpgs <- read_lines(cpg_file)

plp <- read_tsv(plp_file) %>%
  dplyr::filter(gene_symbol_vep %in% cpgs) %>%
  dplyr::rename(Hugo_Symbol = gene_symbol_vep) %>%
  bind_rows(read_tsv(plp_lowVAF_file)) %>%
  arrange(variant_classification_vep) %>%
  distinct(Kids_First_Biospecimen_ID_normal,
                Hugo_Symbol,
           .keep_all = TRUE) %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal,
                Hugo_Symbol,
                variant_classification_vep)

plp_sv <- read_tsv(plp_sv_file) %>%
  dplyr::rename(Hugo_Symbol = Hugo_Symbol_cpg) %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal,
                Hugo_Symbol)
```

Merge all P/LP variants and group into broad variant classification categories

```{r}
plp_cpg <- plp %>%
  bind_rows(plp_sv) %>%
  # filter out variants with multiple classifications that may affect gene expression
  dplyr::filter(!grepl("frameshift_variant&splice|missense_variant&splice|stop_gained&splice|stop_gained&frameshift",
                       variant_classification_vep)) %>%
  dplyr::mutate(variant_classification = case_when(
    is.na(variant_classification_vep) ~ "structural\n variant",
    grepl("frameshift", variant_classification_vep) ~ "frameshift",
    grepl("stop", variant_classification_vep) ~ "stop\n gained",
    grepl("missense", variant_classification_vep) ~ "missense",
    grepl("splice", variant_classification_vep) ~ "splice",
    TRUE ~ "other"
  ))
```

Load gene expression z-scores and append P/LP variant data

```{r}
expr_df <- read_tsv(expr_file)

plp_expr <- plp_cpg %>%
  left_join(expr_df %>% dplyr::select(Kids_First_Biospecimen_ID_normal,
                                   Hugo_Symbol,
                                   expr_zscore),
            by = c("Kids_First_Biospecimen_ID_normal", "Hugo_Symbol")) %>%
  dplyr::mutate(variant_classification = fct_relevel(variant_classification,
                                                     c("structural\n variant",
                                                       "frameshift",
                                                       "splice", "stop\n gained",
                                                       "missense", "other")))
```

plot expression z-scores by variant classification, using splice variant group as a reference

```{r}
plp_expr %>%
  ggplot(aes(x = variant_classification, y = expr_zscore)) + 
    geom_jitter(aes(fill = variant_classification), shape = 21, width = 0.2, size = 2.5, alpha = 0.85,
                show.legend = FALSE) +
    geom_boxplot(alpha = 0.05, outlier.shape = NA) +
    stat_compare_means(method = "wilcox",
                       label = "p.signif",
                       step.increase = 0.1,
                       comparisons = list(c("splice", "structural\n variant"),
                                          c("splice", "frameshift"),
                                          c("splice", "stop\n gained"),
                                          c("splice", "missense"),
                                          c("splice", "other")),
                       method.args = list(alternative = "two.sided"),
                       ) +
    labs(x = "Germline variant classification",
         y = "TPM z-score") +
    ylim(c(-3, 6)) +
    scale_fill_npg() + 
    theme_Publication()

ggsave(file.path(plot_dir,
                 "tpm-zscores-by-variant-classification.pdf"),
       width = 6, height = 4)
```
Load protein abundance z-score file and append to cpg germline variant df

```{r}
prot <- read_tsv(prot_file)

plp_prot <- plp_cpg %>%
  left_join(prot %>% dplyr::select(Kids_First_Biospecimen_ID_normal,
                                   GeneSymbol,
                                   expr_zscore_proteo),
            by = c("Kids_First_Biospecimen_ID_normal", "Hugo_Symbol" = "GeneSymbol")) %>%
  dplyr::mutate(variant_classification = fct_relevel(variant_classification,
                                                     c("structural\n variant",
                                                       "frameshift",
                                                       "splice", "stop\n gained",
                                                       "missense", "other")))
```

Plot CPG P/LP carrier z-scores by germline variant classification

```{r}
plp_prot %>%
  ggplot(aes(x = variant_classification, y = expr_zscore_proteo)) + 
    geom_jitter(aes(fill = variant_classification), shape = 21, width = 0.2, size = 2.5, alpha = 0.85,
                show.legend = FALSE) +
    geom_boxplot(alpha = 0.05, outlier.shape = NA) +
    stat_compare_means(method = "wilcox",
                       label = "p.signif",
                       step.increase = 0.1,
                       comparisons = list(c("splice", "structural\n variant"),
                                          c("splice", "frameshift"),
                                          c("splice", "stop\n gained"),
                                          c("splice", "missense")),
                       method.args = list(alternative = "two.sided"),
                       ) +
    labs(x = "Germline variant classification",
         y = "Protein abundance z-score") +
  #  ylim(c(-3, 6)) +
    scale_fill_npg() + 
    theme_Publication()

ggsave(file.path(plot_dir,
                 "total-protein-zscores-by-variant-classification.pdf"),
       width = 6, height = 4)
```

Load P/LP variant-proximal alternative splicing events df and append expression z-scores.
NOTE: we will re-calculate minimum distance between P/LP variant and splice event here to include distance direction, rather than absolute distance

```{r}
splice_events <- read_tsv(file.path(results_dir,
                                    "plp-variant-proximal-splicing-event-psi.tsv")) %>%
  left_join(expr_df %>% dplyr::select(Kids_First_Biospecimen_ID_rna, 
                                   Hugo_Symbol, expr_zscore)) %>%
  left_join(prot %>% dplyr::select(Kids_First_Biospecimen_ID_normal,
                                   GeneSymbol,
                                   expr_zscore_proteo),
            by = c("Kids_First_Biospecimen_ID_normal", "Hugo_Symbol" = "GeneSymbol")) %>%
  dplyr::mutate(region_start_dist = plp_variant_position - region_start,
                region_end_dist = plp_variant_position - region_end,
                upstream_exon_start_dist = plp_variant_position - upstream_exon_start,
                upstream_exon_end_dist = plp_variant_position - upstream_exon_end,
                downstream_exon_start_dist = plp_variant_position - downstream_exon_start,
                downstream_exon_end_dist = plp_variant_position - downstream_exon_end) %>%
  dplyr::rename(abs_min_dist = min_dist) %>%
  dplyr::mutate(min_dist = pmin(region_start_dist, region_end_dist,
                                upstream_exon_start_dist, upstream_exon_end_dist,
                                downstream_exon_start_dist, downstream_exon_end_dist)) %>%
  dplyr::mutate(max_dist = pmax(region_start_dist, region_end_dist,
                                upstream_exon_start_dist, upstream_exon_end_dist,
                                downstream_exon_start_dist, downstream_exon_end_dist))
```
Calculate pearson correlation coefficients between TPM z-scores and PSI z-scores in P/LP carrirs with frameshift, stop gained, and proximal splice variants
NOTE: we will remove carriers for which the P/LP variant is far downstream (>1kb) of the splice event. 

```{r}
splice_events <- splice_events %>%
    dplyr::mutate(variant_classification = case_when(
      variant_classification == "stop_gained&frameshift_variant" ~ "frameshift_variant",
      TRUE ~ variant_classification
    )) %>%
    dplyr::mutate(plot_case = case_when(
    splicing_case == "SE" ~ "Single Exon", 
    splicing_case == "RI" ~ "Retained Intron",
    splicing_case %in% c("A3SS", "A5SS") ~ "Alternative\n Splice Site",
  )) %>%
  dplyr::mutate(variant_classification = str_replace_all(variant_classification,
                                                     "_", " "))

splice_events %>%
  dplyr::filter(variant_classification %in% c("frameshift variant", 
                                              "stop gained"),
                ((strand == "+" & min_dist < 1000) | (strand == "-" & max_dist > -1000))) %>%
  ggplot(aes(x = sample_PSI_zscore, y = expr_zscore)) +
         geom_point() +
         stat_smooth(method = "lm", 
                     formula = y ~ x, 
                     geom = "smooth", 
                     colour = "red",
                     fill = "pink",
                     linetype="dashed") +
         labs(x = "PSI z-score",
              y = "TPM z-score") + 
         stat_cor(method = "pearson",
                  label.x.npc = "left",
                  label.y.npc = "top", size = 3) +
          facet_wrap(~plot_case, nrow = 1,
                     scale = "free_x") + 
  ylim(c(NA, 5)) +
         theme_Publication()

ggsave(file.path(plot_dir,
       "tpm-psi-correlations-by-splicing-case.pdf"),
       width = 7, height = 3)
```

Re-plot TPM vs. PSI z-scores by splicing case and P/LP variant type (frameshift, stop gained, and splice)

```{r}
# retained introns
splice_events %>%
  dplyr::filter(splicing_case == "RI",
                variant_classification %in% c("frameshift variant", 
                                              "stop gained", 
                                              "splice variant"),
                ((strand == "+" & min_dist < 1000) | (strand == "-" & max_dist > -1000))) %>%
  plot_tpm_vs_psi(., xlab = "Retained Intron\nPSI z-score")
  
ggsave(file.path(plot_dir,
       "tpm-retained-intron-psi-correlations-by-variant-type.pdf"),
       width = 3, height = 7)

# single exon
splice_events %>%
  dplyr::filter(splicing_case == "SE",
                variant_classification %in% c("frameshift variant", 
                                              "stop gained", 
                                              "splice variant"),
                ((strand == "+" & min_dist < 1000) | (strand == "-" & max_dist > -1000))) %>%
  plot_tpm_vs_psi(., xlab = "Skipped Exon\nPSI z-score")

ggsave(file.path(plot_dir,
       "tpm-single-exon-psi-correlations-by-variant-type.pdf"),
       width = 3, height = 7)

# lat splice site
splice_events %>%
  dplyr::filter(splicing_case %in% c("A5SS", "A3SS"),
                variant_classification %in% c("frameshift variant", 
                                              "stop gained", 
                                              "splice variant"),
                ((strand == "+" & min_dist < 1000) | (strand == "-" & max_dist > -1000))) %>%
  plot_tpm_vs_psi(., xlab = "Alt. Splice Site\nPSI z-score")

ggsave(file.path(plot_dir,
       "tpm-altss-psi-correlations-by-variant-type.pdf"),
       width = 3, height = 7)
```

```{r}
sessionInfo()
```