# Identify germline P/LP variant-associated alternative splicing events in PBTA germline cohort

This module queries alternative splicing event RMATs file to identify germline P/LP variant-proximal alternative splicing events in matched tumors, and assesses associations between splice variants, alterantive splicing levels, and gene expression.

## Usage

`bash run_module.sh`

## Folder contents

0. `00-filter-rmats.R` filters PBTA rMATs file for samples in cohort and CPGs
1. `01-calculate-PSI-zscores.Rmd` calculates PSI z-scores of alternative splicing events proximal to germline P/LP variants in P/LP carriers

## Analysis module directory structure

```
.
├── 00-filter-rmats.R
├── 01-calculate-PSI-zscores.Rmd
├── 01-calculate-PSI-zscores.nb.html
├── README.md
├── input
│   └── gtf-annotated-splice-events.tsv
├── plots
│   └── alternative-splicing-PSI-zscores-by-variant-type.pdf
├── results
│   ├── plp-variant-proximal-splicing-event-psi.tsv
│   └── splice-events-rmats-cpgs.tsv.gz
└── run_module.sh
```

## Notes

The input file `gtf-annotated-splice-events.tsv` was generated by running rMATs turbo v4.3.0 using an empty bam file (file path stored in `empty_bam.txt`) and gencode v39 annotation as follows: 

```
python rmats.py -b1 empty_bam.txt -gtf gencode.v39.primary_assembly.annotation.gtf.gz --tmp tmp --od rmats_results --statoff 
```
The resulting `fromGTF.[AS_Event].txt` files were subsequently merged to create an index of all gtf-annotated splice events in gencode v39.
