---
title: "Query germline samples for DNA repair variants"
output: 
  html_document:
  toc: TRUE
toc_float: TRUE
author: Ryan Corbett, Jo Lynne Rokita
date: "2023"
---
  
```{r load libraries and set directories}
suppressPackageStartupMessages({
  library(data.table)
  library(tidyverse)
  library(ComplexHeatmap)
  library(circlize)
})

root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "dna-repair-variant-summary")
results_dir <- file.path(analysis_dir, "results")
input_dir <- file.path(analysis_dir, "input")
plot_dir <- file.path(analysis_dir, "plots")

# source figure theme
source(file.path(root_dir, "figures", "theme.R"))
```

Set histology, PLP variant, and gene lists

```{r}
histologies_file <- file.path(root_dir, "analyses", "collapse-tumor-histologies",
                              "results","germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv")

plp_file <- file.path(data_dir, 
                      "pbta-merged-plp-variants-autogvp-abridged.tsv")

plp_sv_file <- file.path(data_dir,
                         "pbta_germline_svs.tsv")

cpg_file <- file.path(root_dir, "analyses", "oncokb-annotation", "input", "cpg.txt")

```

wrangle data

```{r load germline, hist, cpg}

hist <- read_tsv(histologies_file,show_col_types = FALSE)

germline <- read_tsv(plp_file, show_col_types = FALSE) %>%
  dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist$Kids_First_Biospecimen_ID_normal)

germline_sv <- read_tsv(plp_sv_file, show_col_types = FALSE) %>%
  dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist$Kids_First_Biospecimen_ID_normal)

cpgs <- read_lines(cpg_file)

```

Using DNA repair gene lists obtained from Knijnenburg et al. and Broad GSEA, identify patients with germline PLP variants in DNA repair genes and save results to separate outputs

```{r}
# Define repair list directories and results directories
repair_gene_source <- c("Knijnenburg_paper", "broad")
output_names <- c("Knijnenburg_repair_genes", "Broad_GO_repair_genes")

# loop through two gene list inputs
for (i in 1:length(repair_gene_source)){
  
  # define gene list-specific results and plots dir
  results_dir <- file.path(analysis_dir, "results", output_names[i])
  plot_dir <- file.path(analysis_dir, "plots", output_names[i])
  
  # read in full DNA repair gene list and specific pathway gene lists
  repair_genes <- read_lines(file.path(input_dir, repair_gene_source[i], "dna_repair_all.txt"))
  ber_genes <- read_lines(file.path(input_dir, repair_gene_source[i], "base_excision_repair.txt"))
  hr_genes <- read_lines(file.path(input_dir, repair_gene_source[i], "homologous_recombination.txt"))
  mmr_genes <- read_lines(file.path(input_dir, repair_gene_source[i], "mismatch_repair.txt"))
  ner_genes <- read_lines(file.path(input_dir, repair_gene_source[i], "nucleotide_excision_repair.txt"))
  nhej_genes <- read_lines(file.path(input_dir, repair_gene_source[i], "nonhomologous_end_joining.txt"))

  # filter germline PLP variants for DNA repair genes 
  repair_germline <- germline %>%
    filter(gene_symbol_vep %in% repair_genes & gene_symbol_vep %in% cpgs) %>%
  
  # add columns indicating if gene is involved in specific DNA repair pathways
    mutate(BER = case_when(
      gene_symbol_vep %in% ber_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(HR = case_when(
      gene_symbol_vep %in% hr_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(MMR = case_when(
      gene_symbol_vep %in% mmr_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(NER = case_when(
      gene_symbol_vep %in% ner_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(NHEJ = case_when(
      gene_symbol_vep %in% nhej_genes ~ "Yes",
      TRUE ~ "No")) %>%
    
    # write output to file
    readr::write_tsv(file.path(results_dir, glue::glue("dna_repair_germline_variant_plp_{repair_gene_source[i]}.tsv")))
  
  repair_germline_sv <- germline_sv %>%
    dplyr::rename(gene_symbol_vep = Hugo_Symbol_cpg) %>%
    filter(gene_symbol_vep %in% repair_genes & gene_symbol_vep %in% cpgs) %>%
  
  # add columns indicating if gene is involved in specific DNA repair pathways
    mutate(BER = case_when(
      gene_symbol_vep %in% ber_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(HR = case_when(
      gene_symbol_vep %in% hr_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(MMR = case_when(
      gene_symbol_vep %in% mmr_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(NER = case_when(
      gene_symbol_vep %in% ner_genes ~ "Yes",
      TRUE ~ "No")) %>%
    mutate(NHEJ = case_when(
      gene_symbol_vep %in% nhej_genes ~ "Yes",
      TRUE ~ "No")) %>%
    
    # write output to file
    readr::write_tsv(file.path(results_dir, glue::glue("dna_repair_germline_sv_plp_{repair_gene_source[i]}.tsv")))

  # define `plot_df` for plotting DNA repair pathway PLP variants by plot group 
  plot_df <- repair_germline %>%
    filter(BER == "Yes") %>%
    bind_rows(repair_germline[repair_germline$HR == "Yes",]) %>%
    bind_rows(repair_germline[repair_germline$MMR == "Yes",]) %>%
    bind_rows(repair_germline[repair_germline$NER == "Yes",]) %>%
    bind_rows(repair_germline[repair_germline$NHEJ == "Yes",])
  
  plot_df <- plot_df %>%
    bind_rows(repair_germline[!repair_germline$Kids_First_Biospecimen_ID_normal %in% plot_df$Kids_First_Biospecimen_ID_normal,]) %>%
    mutate(class = c(rep("BER", nrow(repair_germline[repair_germline$BER == "Yes",])),
                     rep("HR", nrow(repair_germline[repair_germline$HR == "Yes",])),
                     rep("MMR", nrow(repair_germline[repair_germline$MMR == "Yes",])),
                     rep("NER", nrow(repair_germline[repair_germline$NER == "Yes",])),
                     rep("NHEJ", nrow(repair_germline[repair_germline$NHEJ == "Yes",])),
                     rep("Other", nrow(repair_germline[!repair_germline$Kids_First_Biospecimen_ID_normal %in% plot_df$Kids_First_Biospecimen_ID_normal,]))
                     )
           )
  
  plot_sv_df <- repair_germline_sv %>%
    filter(BER == "Yes") %>%
    bind_rows(repair_germline_sv[repair_germline_sv$HR == "Yes",]) %>%
    bind_rows(repair_germline_sv[repair_germline_sv$MMR == "Yes",]) %>%
    bind_rows(repair_germline_sv[repair_germline_sv$NER == "Yes",]) %>%
    bind_rows(repair_germline_sv[repair_germline_sv$NHEJ == "Yes",])
  
  plot_sv_df <- plot_sv_df %>%
    bind_rows(repair_germline_sv[!repair_germline_sv$Kids_First_Biospecimen_ID_normal %in% plot_sv_df$Kids_First_Biospecimen_ID_normal,]) %>%
    mutate(class = c(rep("BER", nrow(repair_germline_sv[repair_germline_sv$BER == "Yes",])),
                     rep("HR", nrow(repair_germline_sv[repair_germline_sv$HR == "Yes",])),
                     rep("MMR", nrow(repair_germline_sv[repair_germline_sv$MMR == "Yes",])),
                     rep("NER", nrow(repair_germline_sv[repair_germline_sv$NER == "Yes",])),
                     rep("NHEJ", nrow(repair_germline_sv[repair_germline_sv$NHEJ == "Yes",])),
                     rep("Other", nrow(repair_germline_sv[!repair_germline_sv$Kids_First_Biospecimen_ID_normal %in% plot_sv_df$Kids_First_Biospecimen_ID_normal,]))
                     )
           )
  
  
  plot_df <- plot_df %>%
    dplyr::select(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, class) %>%
    bind_rows(plot_sv_df %>% dplyr::select(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, class)) %>%
    distinct(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, class) %>%
    # add plot group 
    left_join(hist[,c("Kids_First_Biospecimen_ID_normal", "plot_group", "plot_group_hex")], by = "Kids_First_Biospecimen_ID_normal") %>%
    # retain only one instace of pathway P/LP variant per pt
    distinct(Kids_First_Biospecimen_ID_normal, class, .keep_all = T) %>%
    count(class, plot_group, plot_group_hex)
  
    # define x axis order  
    level_order <- plot_df %>%
      group_by(class) %>%
      summarise(total_n = sum(n)) %>%
      arrange(total_n) %>%
      pull(class)
    
    plot_df <- plot_df %>%
      dplyr::mutate(class = factor(class)) %>%
      dplyr::mutate(class = fct_relevel(class, level_order))
    
    # define plot group pallete 
    plot_group_palette <- plot_df$plot_group_hex
    names(plot_group_palette) <- plot_df$plot_group

    # Create barplot of DNA repair pathway germline variant prevalence by plot group
    ggplot(plot_df, aes(x = class, y = n, fill = plot_group)) +
    geom_col(col = "black", size = 0.4) +
    labs(y = "Number of patients with P/LP variant", x = "Germline variant DNA repair class", fill = "Tumor diagnosis") +
    scale_fill_manual(values = plot_group_palette) + 
    theme_Publication()
    
    ggsave(
    file.path(plot_dir, paste0("dna_repair_samples_by_plot_group_", repair_gene_source[i], ".pdf")), 
    width = 24, height = 16, units = "cm")
    
    # Add columns indicate presence/absence of DNA repair P-LP variants to histologies file
    hist_repair <- hist %>%
      dplyr::mutate(`All Repair` = case_when(
        Kids_First_Biospecimen_ID_normal %in% c(repair_germline$Kids_First_Biospecimen_ID_normal,
                                                repair_germline_sv$Kids_First_Biospecimen_ID_normal) ~ "Yes",
        TRUE ~ "No"
      )) %>%
      dplyr::mutate(BER = case_when(
        Kids_First_Biospecimen_ID_normal %in% c(repair_germline$Kids_First_Biospecimen_ID_normal[repair_germline$BER == "Yes"],
                                                repair_germline_sv$Kids_First_Biospecimen_ID_normal[repair_germline_sv$BER == "Yes"]) ~ "Yes",
        TRUE ~ "No"
      )) %>%
      dplyr::mutate(HR = case_when(
        Kids_First_Biospecimen_ID_normal %in% c(repair_germline$Kids_First_Biospecimen_ID_normal[repair_germline$HR == "Yes"],
                                                repair_germline_sv$Kids_First_Biospecimen_ID_normal[repair_germline_sv$HR == "Yes"]) ~ "Yes",
        TRUE ~ "No"
      )) %>%
      dplyr::mutate(MMR = case_when(
        Kids_First_Biospecimen_ID_normal %in% c(repair_germline$Kids_First_Biospecimen_ID_normal[repair_germline$MMR == "Yes"],
                                                repair_germline_sv$Kids_First_Biospecimen_ID_normal[repair_germline_sv$MMR == "Yes"]) ~ "Yes",
        TRUE ~ "No"
      )) %>%
      dplyr::mutate(NER = case_when(
        Kids_First_Biospecimen_ID_normal %in% c(repair_germline$Kids_First_Biospecimen_ID_normal[repair_germline$NER == "Yes"],
                                                repair_germline_sv$Kids_First_Biospecimen_ID_normal[repair_germline_sv$NER == "Yes"]) ~ "Yes",
        TRUE ~ "No"
      )) %>%
      dplyr::mutate(NHEJ = case_when(
        Kids_First_Biospecimen_ID_normal %in% c(repair_germline$Kids_First_Biospecimen_ID_normal[repair_germline$NHEJ == "Yes"],
                                                repair_germline_sv$Kids_First_Biospecimen_ID_normal[repair_germline_sv$NHEJ == "Yes"]) ~ "Yes",
        TRUE ~ "No"
      )) %>%
      dplyr::mutate(Other = case_when(
        `All Repair` == "Yes" & BER == "No" & HR == "No" & NER == "No" & MMR == "No" & NHEJ == "No" ~ "Yes",
        TRUE ~ "No"
      ))

    # Run DNA repair gene P-LP enrichment by histology group and repair pathway
    
    groups <- unique(hist_repair$plot_group)
    groups <- groups[order(groups)]
    paths <- c("All Repair", "BER", "HR", "MMR", "NER", "NHEJ", "Other")
    
    # Define count, odds ratio and p-value matrices
    
    enr_mat <- matrix(0, length(groups), length(paths),
                      dimnames = list(groups, paths))
    pval_mat <- enr_mat
    ct_mat <- enr_mat
    
    # Run Fisher's exact tests by histology and repair pathway
    for (group in groups) {
      
      for (path in paths){
        
        in_hist_path <- hist_repair %>%
          dplyr::filter(plot_group == group & hist_repair[,path] == "Yes") %>%
          pull(Kids_First_Biospecimen_ID_normal)
        
        in_hist_no_path <- hist_repair %>%
          dplyr::filter(plot_group == group & hist_repair[,path] == "No") %>%
          pull(Kids_First_Biospecimen_ID_normal)
        
        no_hist_in_path <- hist_repair %>%
          dplyr::filter(plot_group != group & hist_repair[,path] == "Yes") %>%
          pull(Kids_First_Biospecimen_ID_normal)
        
        no_hist_no_path <- hist_repair %>%
          dplyr::filter(plot_group != group & hist_repair[,path] == "No") %>%
          pull(Kids_First_Biospecimen_ID_normal)
        
        mat <- matrix(c(length(in_hist_path), length(in_hist_no_path),
                        length(no_hist_in_path), length(no_hist_no_path)),
                      2, 2)
        
        ct_mat[group, path] <- length(in_hist_path)
        enr_mat[group, path] <- fisher.test(mat, alternative = "greater")$estimate
        pval_mat[group, path] <- fisher.test(mat, alternative = "greater")$p.value
      }
    }
    
    # multiple test correction
    fdr_mat <- t(apply(pval_mat, 1, function(x) p.adjust(x, method = "BH")))
  
  # create text matrix of odds ratios and asterisk in cases of statistical significance
  sig_mat <- ifelse(fdr_mat < 0.05 & enr_mat > 1, "*", "")
  
  fill_mat <- matrix(glue::glue("{ct_mat}\n({round(enr_mat, 1)}{sig_mat})"), 
                     length(groups), length(paths))
  
  fill_mat <- ifelse(fill_mat == "0\n(0)", "0", fill_mat)
  
  # Plot enrichment heatmap
  
  col_fun = colorRamp2(c(0, 6), c("white", "orangered"))
  
  pdf(file.path(plot_dir, paste0("dna_repair_enr_by_plot_group_", repair_gene_source[i], ".pdf")),
        width = 6, height = 7)
  
  enr_heatmap <- Heatmap(enr_mat,
                         name = "Odds ratio",
                         cluster_rows = F,
                         cluster_columns = F,
                         rect_gp = gpar(col = "black", lwd = 2),
                         col = col_fun,
                         cell_fun = function(j, i, x, y, width, height, fill) {
                         grid.text(sprintf("%s", fill_mat[i, j]), x, y, gp = gpar(fontsize = 10))
                 })
  
  print(enr_heatmap)

  dev.off()

}
```