---
title: 'GSVA score comparison in HGG tumors'
output: 
  html_document:
  toc: TRUE
toc_float: TRUE
author: Ryan Corbett
date: "2023"
---

This script compares GSVA scores for hallmark and KEGG pathways between PBTA samples with and without DNA repair gene germline PLP variants

Load packages and set directories

```{r load libraries, message = FALSE, warning = FALSE}
library(tidyverse)
#library(data.table)
library(ggplot2)
library(ggsci)
library(ggpubr)

root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "dna-repair-variant-summary")
input_dir <- file.path(analysis_dir, "input")

source(file.path(analysis_dir, "util", "gsva_functions.R"))
source(file.path(root_dir, "figures", "theme.R"))
```

Set file paths

```{r set file paths}

cbtn_histologies_file <- file.path(root_dir, 'analyses', 'collapse-tumor-histologies', 'results',"germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv")

pbta_histologies_file <- file.path(data_dir,"histologies.tsv")

independent_specimens_file <- file.path(data_dir, "independent-specimens.rnaseqpanel.primary-plus.tsv")

gsva_file <- file.path(input_dir, 'pbta-hallmark-gsva-scores.tsv')
kegg_file <- file.path(analysis_dir, "results", "pbta-KEGG-gsva-scores.tsv")

maf_file <- file.path(root_dir, "analyses", "oncokb-annotation", "results", "snv-consensus-plus-hotspots-goi-oncokb.maf.tsv")

```

Read in CBTN and PBTA histologies, gsva results, maf, and cpg files. Filter histologies file for high-grade glioma only

```{r}

hist_cbtn <- read_tsv(cbtn_histologies_file) %>%
  dplyr::filter(grepl("Other high-grade glioma", plot_group),
         grepl("wildtype", molecular_subtype))

hist_pbta <- read_tsv(pbta_histologies_file) 

independent_specimens <- read_tsv(independent_specimens_file)

hallmark_scores <- read_tsv(gsva_file)
hallmark_pathways <- unique(hallmark_scores$hallmark_name)

kegg_scores <- read_tsv(kegg_file)
kegg_pathways <- unique(kegg_scores$kegg_name)

maf <- data.table::fread(maf_file, data.table = F) %>%
  dplyr::filter(Tumor_Sample_Barcode %in% hist_cbtn$Kids_First_Biospecimen_ID_tumor)

cpgs <- read_lines(file.path(root_dir, "analyses", "oncokb-annotation", "input", "cpg.txt"))

```

Pull BS_ids corresponding to RNA-Seq experiments. Append GSVA scores for samples and calculate median scores in instances where there are >1 RNA-seq sample per patient

```{r}

# rna_ids <- hist_pbta %>% 
#   dplyr::filter(experimental_strategy == 'RNA-Seq' & sample_id %in% hist_cbtn$sample_id_tumor) %>%
#   pull(Kids_First_Biospecimen_ID)

rna_ids <- independent_specimens %>%
  pull(Kids_First_Biospecimen_ID)

hallmark_scores <- hallmark_scores %>%
  dplyr::filter(Kids_First_Biospecimen_ID %in% rna_ids) %>%
  left_join(hist_pbta[,c("Kids_First_Biospecimen_ID", "Kids_First_Participant_ID")], by = "Kids_First_Biospecimen_ID")

kegg_scores <- kegg_scores %>%
  dplyr::filter(Kids_First_Biospecimen_ID %in% rna_ids) %>%
  left_join(hist_pbta[,c("Kids_First_Biospecimen_ID", "Kids_First_Participant_ID")], by = "Kids_First_Biospecimen_ID")

```

Run GSVA score comparisons between samples with and without DNA repair germline variants using both DNA repair gene list sources (Knijnenburg et al. and GSEA Broad)

```{r}
# Define repair gene source input directories 
repair_gene_source <- c("Knijnenburg_repair_genes", "Broad_GO_repair_genes")

# Loop through two DNA repair gene lists
for (i in 1:length(repair_gene_source)){
  
  # define gene list-specific results and plot directories
  results_dir <- file.path(analysis_dir, "results", repair_gene_source[i])
  plot_dir <- file.path(analysis_dir, "plots", repair_gene_source[i])
  
  # read in DNA repair gene germline PLP file
  if (repair_gene_source[i] == "Knijnenburg_repair_genes"){
      dna_repair_plp_file <- file.path(results_dir, "dna_repair_germline_variant_plp_Knijnenburg_paper.tsv")
      dna_repair_sv_plp_file <- file.path(results_dir, "dna_repair_germline_sv_plp_Knijnenburg_paper.tsv")
      }else{
        dna_repair_plp_file <- file.path(results_dir, "dna_repair_germline_variant_plp_broad.tsv")
        dna_repair_sv_plp_file <- file.path(results_dir, "dna_repair_germline_sv_plp_broad.tsv")
      }
    
    dna_repair_plp <- read_tsv(dna_repair_plp_file, show_col_types = FALSE)
    dna_repair_sv_plp <- read_tsv(dna_repair_sv_plp_file, show_col_types = FALSE)
    
    dna_repair_plp <- dna_repair_plp %>%
      dplyr::select(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, 
                    BER, HR, MMR, NER, NHEJ) %>%
      bind_rows(dna_repair_sv_plp %>% dplyr::select(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, 
                                                    BER, HR, MMR, NER, NHEJ))
  
  # obtain normal BS_ids for patients with germline PLP variant in: 1) any DNA repair gene, 2) MMR genes, and 3) Other DNA repair genes, excluding MMR
  allRepair_germline <- dna_repair_plp %>%
    dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist_cbtn$Kids_First_Biospecimen_ID_normal) %>%
    distinct(Kids_First_Biospecimen_ID_normal) %>%
    pull(Kids_First_Biospecimen_ID_normal)

  mmr_germline <- dna_repair_plp %>%
    dplyr::filter(MMR == "Yes" & Kids_First_Biospecimen_ID_normal %in% hist_cbtn$Kids_First_Biospecimen_ID_normal) %>%
    distinct(Kids_First_Biospecimen_ID_normal) %>%
    pull(Kids_First_Biospecimen_ID_normal)
  
  brca_germline <- dna_repair_plp %>%
    dplyr::filter(grepl("BRCA|BARD1|BRIP1|ATM|CHEK2|RAD51|PALB2", gene_symbol_vep) & !grepl("RAD51C", gene_symbol_vep)) %>%
    dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist_cbtn$Kids_First_Biospecimen_ID_normal) %>%
    distinct(Kids_First_Biospecimen_ID_normal) %>%
    pull(Kids_First_Biospecimen_ID_normal)

  otherRepair_germline <- dna_repair_plp %>%
    dplyr::filter(!Kids_First_Biospecimen_ID_normal %in% c(mmr_germline, brca_germline) & Kids_First_Biospecimen_ID_normal %in% hist_cbtn$Kids_First_Biospecimen_ID_normal) %>%
    distinct(Kids_First_Biospecimen_ID_normal) %>%
    pull(Kids_First_Biospecimen_ID_normal)
    
   # define list of BS_ids in each group, and names for subsequent plots
  sample_list <- list(allRepair_germline, mmr_germline, brca_germline, otherRepair_germline)
  names(sample_list) <- c("All_repair", "MMR", "BRCA", "Other_repair")
  plot_names <- c("DNA repair", "MMR", "BRCA/BRCA-interacting", "Other DNA repair")
  
  # Loop through sample lists to compare GSVA scores between group of interest and samples without DNA repair germline variants
  for (j in 1:length(sample_list)){
  
    repair_germline <- sample_list[[j]]
    
    ctrl_germline <- hist_cbtn %>%
      dplyr::filter(!Kids_First_Biospecimen_ID_normal %in% repair_germline) %>%
      pull(Kids_First_Biospecimen_ID_normal)
    
    # obtain corresponding RNA-seq tumors BS_IDs for `repair_germline` and `ctrl_germline` samples 
    repair_ids <- hist_pbta %>%
      dplyr::filter(Kids_First_Participant_ID %in% Kids_First_Participant_ID[Kids_First_Biospecimen_ID %in% repair_germline]) %>%
      dplyr::filter(Kids_First_Biospecimen_ID %in% rna_ids) %>%
      pull(Kids_First_Participant_ID) %>%
      unique()

    ctrl_ids <- hist_pbta %>%
      dplyr::filter(Kids_First_Participant_ID %in% Kids_First_Participant_ID[Kids_First_Biospecimen_ID %in% ctrl_germline]) %>%
      dplyr::filter(Kids_First_Biospecimen_ID %in% rna_ids) %>%
      pull(Kids_First_Participant_ID) %>%
      unique()
    
    # define data frame to store GSVA scores, differences between groups, and corresponding test statistic p-values 
    gsva_res <- data.frame(hallmark = hallmark_pathways,
                           repair_score = rep(0, length(hallmark_pathways)),
                           ctrl_score = rep(0, length(hallmark_pathways)),
                           score_diff = rep(0, length(hallmark_pathways)),
                           wilcox_stat = rep(0, length(hallmark_pathways)),
                           pvalue = rep(0, length(hallmark_pathways)))

    # loop through hallmark pathways to obtain score means and differences between groups
    for (k in 1:length(hallmark_pathways)){
      repair_scores <- hallmark_scores[hallmark_scores$hallmark_name == hallmark_pathways[k] & hallmark_scores$`Kids_First_Participant_ID` %in% repair_ids,]$gsea_score
      ctrl_scores <- hallmark_scores[hallmark_scores$hallmark_name == hallmark_pathways[k] & hallmark_scores$`Kids_First_Participant_ID` %in% ctrl_ids,]$gsea_score
      gsva_res$repair_score[k] <- mean(repair_scores)
      gsva_res$ctrl_score[k] <- mean(ctrl_scores)
      gsva_res$score_diff[k] <- mean(repair_scores) - mean(ctrl_scores)
      gsva_res$wilcox_stat[k] <- ifelse(!is.nan(mean(repair_scores) - mean(ctrl_scores)),
                                       wilcox.test(repair_scores, ctrl_scores, alternative = 'two.sided')$statistic,
                                       NA)
      gsva_res$pvalue[k] <- ifelse(!is.nan(mean(repair_scores) - mean(ctrl_scores)),
                                       wilcox.test(repair_scores, ctrl_scores, alternative = 'two.sided')$p.value,
                                       NA)
    }
    
    # write to output
    write_tsv(gsva_res, file.path(results_dir, paste0("gsva_hallmark_score_diff_", 
                                                          names(sample_list)[j], 
                                                          "_vs_other.tsv")))
    
    # create barplot of score differences between groups
    gsva_res <- gsva_res %>%
      dplyr::filter(!is.na(pvalue) & abs(score_diff) > 0.1)
    
    gsva_barplot(gsva_res, "score_diff", "hallmark", plot_names[j])
      

    ggsave(
      file.path(plot_dir, paste0("HALLMARK_gsva_diff_", names(sample_list)[j], "_vs_other.tiff")), 
      width = 15, height = 10, units = "cm")
    
    #define data frame to store KEGG GSVA scores, differences between groups, and corresponding test statistic p-values
    kegg_res <- data.frame(pathway = kegg_pathways,
                           repair_score = rep(0, length(kegg_pathways)),
                           ctrl_score = rep(0, length(kegg_pathways)),
                           score_diff = rep(0, length(kegg_pathways)),
                           wilcox_stat = rep(0, length(kegg_pathways)),
                           pvalue = rep(0, length(kegg_pathways)))

    # loop through KEGG pathways to obtain score means and differences between groups
    for (k in 1:length(kegg_pathways)){
      repair_scores <- kegg_scores[kegg_scores$kegg_name == kegg_pathways[k] & kegg_scores$`Kids_First_Participant_ID` %in% repair_ids,]$gsea_score
      ctrl_scores <- kegg_scores[kegg_scores$kegg_name == kegg_pathways[k] & kegg_scores$`Kids_First_Participant_ID` %in% ctrl_ids,]$gsea_score
      kegg_res$repair_score[k] <- mean(repair_scores)
      kegg_res$ctrl_score[k] <- mean(ctrl_scores)
      kegg_res$score_diff[k] <- mean(repair_scores) - mean(ctrl_scores)
      kegg_res$wilcox_stat[k] <- ifelse(!is.nan(mean(repair_scores) - mean(ctrl_scores)),
                                       wilcox.test(repair_scores, ctrl_scores, alternative = 'two.sided')$statistic,
                                       NA)
      kegg_res$pvalue[k] <- ifelse(!is.nan(mean(repair_scores) - mean(ctrl_scores)),
                                       wilcox.test(repair_scores, ctrl_scores, alternative = 'two.sided')$p.value,
                                       NA)
    }

    # write to output
    write_tsv(kegg_res, file.path(results_dir, paste0("gsva_kegg_score_diff_",
                                                          names(sample_list)[j],
                                                          "_vs_other.tsv")))

    # create barplot of score differences between groups
    kegg_res <- kegg_res %>%
      dplyr::filter(!is.na(pvalue) & abs(score_diff) > 0.175 & !grepl("GLYCOS", pathway))

    gsva_barplot(kegg_res, "score_diff", "pathway", plot_names[j])

    ggsave(
      file.path(plot_dir, paste0("KEGG_gsva_diff_", names(sample_list)[j], "_vs_other.tiff")),
      width = 15, height = 10, units = "cm")

    }
}

```

Plot scores of Hallmark pathways of interest by group (MMR germline variant samples, other DNA repair germline variant samples, no DNA repair germline variant samples)

```{r}
# Define hallmark pathways to plot
hallmark_to_plot <- c("HALLMARK_COAGULATION", "HALLMARK_DNA_REPAIR", "HALLMARK_COMPLEMENT",
                      "HALLMARK_KRAS_SIGNALING_UP")
kegg_to_plot <- c("KEGG_LYSOSOME", "KEGG_N_GLYCAN_BIOSYNTHESIS",
                  "KEGG_FOCAL_ADHESION", "KEGG_PRION_DISEASES")

# create plots for each DNA repair gene list source
repair_gene_source <- c("Knijnenburg_repair_genes", "Broad_GO_repair_genes")
  
# loop through repair gene list sources
for (i in 1:length(repair_gene_source)){
  
  # define gene list-specific results and plots dir
  results_dir <- file.path(analysis_dir, "results", repair_gene_source[i])
  plot_dir <- file.path(analysis_dir, "plots", repair_gene_source[i])
  
  # read in DNA repair gene germline PLP file
if (repair_gene_source[i] == "Knijnenburg_repair_genes"){
    dna_repair_plp_file <- file.path(results_dir, "dna_repair_germline_variant_plp_Knijnenburg_paper.tsv")
    dna_repair_sv_plp_file <- file.path(results_dir, "dna_repair_germline_sv_plp_Knijnenburg_paper.tsv")
    }else{
      dna_repair_plp_file <- file.path(results_dir, "dna_repair_germline_variant_plp_broad.tsv")
      dna_repair_sv_plp_file <- file.path(results_dir, "dna_repair_germline_sv_plp_broad.tsv")
    }
  
  dna_repair_plp <- read_tsv(dna_repair_plp_file, show_col_types = FALSE)
  dna_repair_sv_plp <- read_tsv(dna_repair_sv_plp_file, show_col_types = FALSE)
  
  dna_repair_plp <- dna_repair_plp %>%
    dplyr::select(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, 
                  BER, HR, MMR, NER, NHEJ) %>%
    bind_rows(dna_repair_sv_plp %>% dplyr::select(Kids_First_Biospecimen_ID_normal, gene_symbol_vep, 
                                                  BER, HR, MMR, NER, NHEJ))

  # Define BS_ids for normal and corresponding tumor RNA-seq IDs for each sample group 
  mmr_germline <- dna_repair_plp %>%
    dplyr::filter(MMR == "Yes") %>%
    dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist_cbtn$Kids_First_Biospecimen_ID_normal) %>%
    distinct(Kids_First_Biospecimen_ID_normal) %>%
    dplyr::pull(Kids_First_Biospecimen_ID_normal)
  
  brca_germline <- dna_repair_plp %>%
    dplyr::filter(grepl("BRCA|BARD1|BRIP1|ATM|CHEK2|RAD51|PALB2", gene_symbol_vep) & !grepl("RAD51C", gene_symbol_vep)) %>%
    dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist_cbtn$Kids_First_Biospecimen_ID_normal) %>%
    distinct(Kids_First_Biospecimen_ID_normal) %>%
    dplyr::pull(Kids_First_Biospecimen_ID_normal)
  
  otherRepair_germline <- dna_repair_plp %>%
    dplyr::filter(!Kids_First_Biospecimen_ID_normal %in% c(mmr_germline, brca_germline) & Kids_First_Biospecimen_ID_normal %in% hist_cbtn$Kids_First_Biospecimen_ID_normal) %>%
    dplyr::pull(Kids_First_Biospecimen_ID_normal)
  
  ctrl_germline <- hist_cbtn %>%
    dplyr::filter(!Kids_First_Biospecimen_ID_normal %in% dna_repair_plp$Kids_First_Biospecimen_ID_normal) %>%
    dplyr::pull(Kids_First_Biospecimen_ID_normal)
  
  mmr_ids <- hist_pbta %>%
    dplyr::filter(Kids_First_Participant_ID %in% Kids_First_Participant_ID[Kids_First_Biospecimen_ID %in% mmr_germline]) %>%
    dplyr::filter(sample_type == 'Tumor' & experimental_strategy == 'RNA-Seq') %>%
    dplyr::pull(Kids_First_Participant_ID) %>%
    unique()
  
  brca_ids <- hist_pbta %>%
    dplyr::filter(Kids_First_Participant_ID %in% Kids_First_Participant_ID[Kids_First_Biospecimen_ID %in% brca_germline]) %>%
    dplyr::filter(sample_type == 'Tumor' & experimental_strategy == 'RNA-Seq') %>%
    dplyr::pull(Kids_First_Participant_ID) %>%
    unique()
  
  otherRepair_ids <- hist_pbta %>%
    dplyr::filter(Kids_First_Participant_ID %in% Kids_First_Participant_ID[Kids_First_Biospecimen_ID %in% otherRepair_germline]) %>%
    dplyr::filter(sample_type == 'Tumor' & experimental_strategy == 'RNA-Seq') %>%
    dplyr::pull(Kids_First_Participant_ID) %>%
    unique()

  ctrl_ids <- hist_pbta %>%
    dplyr::filter(Kids_First_Participant_ID %in% Kids_First_Participant_ID[Kids_First_Biospecimen_ID %in% ctrl_germline]) %>%
    dplyr::filter(sample_type == 'Tumor' & experimental_strategy == 'RNA-Seq') %>%
    dplyr::pull(Kids_First_Participant_ID) %>%
    unique()
    
  # Loop through hallmark pathways of interest to generate violin plots by group
  for (j in 1:length(hallmark_to_plot)){
    
    # Create data frame for plotting
    score_df <- hallmark_scores %>%
      dplyr::filter(hallmark_name == hallmark_to_plot[j] & Kids_First_Participant_ID %in% c(mmr_ids, brca_ids, otherRepair_ids, ctrl_ids)) %>%
      dplyr::mutate(germline_variant = case_when(
        Kids_First_Participant_ID %in% mmr_ids ~ "MMR",
        Kids_First_Participant_ID %in% brca_ids ~ "BRCA/BRCA-interacting",
        Kids_First_Participant_ID %in% otherRepair_ids ~ "Other repair",
        TRUE ~ "Non-repair"
      )) %>%
      dplyr::mutate(germline_variant = factor(germline_variant)) %>% 
      dplyr::mutate(germline_variant = fct_relevel(germline_variant,
                                            c("MMR", "BRCA/BRCA-interacting", "Other repair", "Non-repair"))
             )
    
    gsva_violin(score_df, "germline_variant", "gsea_score", hallmark_to_plot[j])
        
      ggsave(
        file.path(plot_dir, paste0(hallmark_to_plot[j], "_gsea_score_byGroup_violinPlot.tiff")), 
        width = 14, height = 11, units = "cm")
               
  }
  
  # Create summary table of samples with high (>3rd quartile) DNA repair GSVA scores
  collapsed_germline <- dna_repair_plp %>%
    dplyr::filter(MMR == "Yes" & gene_symbol_vep %in% cpgs) %>%
    group_by(Kids_First_Biospecimen_ID_normal) %>%
    summarise(Hugo_Symbol_germline = str_c(sort(unique(gene_symbol_vep)), collapse = "; "))

  collapsed_onco <- maf %>%
    dplyr::filter(Hugo_Symbol %in% cpgs) %>%
    dplyr::rename(Kids_First_Biospecimen_ID_normal = Matched_Norm_Sample_Barcode,
                  Kids_First_Biospecimen_ID_tumor = Tumor_Sample_Barcode,
                  Hugo_Symbol_somatic = Hugo_Symbol) %>%
    group_by(Kids_First_Biospecimen_ID_normal) %>%
    summarise(Hugo_Symbol_somatic = str_c(sort(unique(Hugo_Symbol_somatic)), collapse = "; "))

  hallmark_scores %>%
    dplyr::filter(hallmark_name == "HALLMARK_DNA_REPAIR" & Kids_First_Participant_ID %in% hist_cbtn$Kids_First_Participant_ID) %>%
    dplyr::filter(gsea_score > 0.4) %>%
    left_join(hist_cbtn[,c("Kids_First_Participant_ID", "Kids_First_Biospecimen_ID_normal",
                          "Kids_First_Biospecimen_ID_tumor", "plot_group", "molecular_subtype")]) %>%
        left_join(collapsed_germline) %>%
        left_join(collapsed_onco) %>%
        dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor,
               plot_group, molecular_subtype, gsea_score,
               Hugo_Symbol_germline, Hugo_Symbol_somatic) %>%
        write_tsv(file.path(results_dir, "PBTA_hgg_samples_with_DNArepair_gsva.tsv"))
  
  for (j in 1:length(kegg_to_plot)){
    
    # Create data frame for plotting
    score_df <- kegg_scores %>%
      dplyr::filter(kegg_name == kegg_to_plot[j] & Kids_First_Participant_ID %in% c(mmr_ids, brca_ids, otherRepair_ids, ctrl_ids)) %>%
      dplyr::mutate(germline_variant = case_when(
        Kids_First_Participant_ID %in% mmr_ids ~ "MMR",
        Kids_First_Participant_ID %in% brca_ids ~ "BRCA/BRCA-interacting",
        Kids_First_Participant_ID %in% otherRepair_ids ~ "Other repair",
        TRUE ~ "Non-repair"
      )) %>%
      dplyr::mutate(germline_variant = factor(germline_variant,
                                       c("MMR", "BRCA/BRCA-interacting", "Other repair", "Non-repair")))
    
    gsva_violin(score_df, "germline_variant", "gsea_score", kegg_to_plot[j])
        
      ggsave(
        file.path(plot_dir, paste0(kegg_to_plot[j], "_gsea_score_byGroup_violinPlot.tiff")), 
        width = 14, height = 11, units = "cm")
               
  }
  
#  Create summary table of samples with high (>3rd quartile) mismatch repair GSVA scores
  collapsed_germline <- dna_repair_plp %>%
    dplyr::filter(MMR == "Yes" & gene_symbol_vep %in% cpgs) %>%
    group_by(Kids_First_Biospecimen_ID_normal) %>%
    summarise(Hugo_Symbol_germline = str_c(sort(unique(gene_symbol_vep)), collapse = "; "))

  collapsed_onco <- maf %>%
    dplyr::filter(Hugo_Symbol %in% cpgs) %>%
    dplyr::rename(Kids_First_Biospecimen_ID_normal = Matched_Norm_Sample_Barcode,
                  Kids_First_Biospecimen_ID_tumor = Tumor_Sample_Barcode,
                  Hugo_Symbol_somatic = Hugo_Symbol) %>%
    group_by(Kids_First_Biospecimen_ID_normal) %>%
    summarise(Hugo_Symbol_somatic = str_c(sort(unique(Hugo_Symbol_somatic)), collapse = "; "))

  kegg_scores %>%
    dplyr::filter(kegg_name == "KEGG_MISMATCH_REPAIR" & Kids_First_Participant_ID %in% hist_cbtn$Kids_First_Participant_ID) %>%
    dplyr::filter(gsea_score > .4) %>%
    left_join(hist_cbtn[,c("Kids_First_Participant_ID", "Kids_First_Biospecimen_ID_normal",
                          "Kids_First_Biospecimen_ID_tumor", "plot_group", "molecular_subtype")]) %>%
        left_join(collapsed_germline) %>%
        left_join(collapsed_onco) %>%
        dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor,
               plot_group, molecular_subtype, gsea_score,
               Hugo_Symbol_germline, Hugo_Symbol_somatic) %>%
        write_tsv(file.path(results_dir, paste0("PBTA_hgg_samples_with_MMR_gsva.tsv")))
      
}

```

```{r echo=TRUE}
sessionInfo()

```
