---
title: "Assess second hits"
output: 
  html_notebook:
  toc: TRUE
toc_float: TRUE
author: Jo Lynne Rokita
params:
  maf_in:
    label: "OncoKB-Annotated MAF input file"
    value: analyses/oncokb-annotation/results/snv-consensus-plus-hotspots-goi-oncokb.maf.tsv
    input: file
  germ_in:
    label: "germline file"
    value: data/pbta-merged-plp-variants-autogvp-abridged.tsv
    input: file
editor_options: 
  chunk_output_type: inline
---
### Set libraries and file paths

```{r load libraries}
# load libraries
library(tidyverse)
library(maftools)

# define directories 
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "two-hits")
results_dir <- file.path(analysis_dir, "results")
data_dir <- file.path(root_dir, "data")
plot_dir <- file.path(analysis_dir, "plots")

# If the input and results directories do not exist, create it
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

# add histologies file so we can subset maf
histologies_file <- file.path(root_dir,"analyses", 
                              "collapse-tumor-histologies", "results", 
                              "germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv")

# remove cell lines from histologies file
opc_hist_file <- file.path(data_dir, "histologies.tsv")

plp_sv_file <- file.path(data_dir, "pbta_germline_svs.tsv")

# goi file
cpg_file <- file.path(root_dir, "analyses", "oncokb-annotation", "input", "cpg.txt")
```

### Load oncokb annotated maf and explore data

```{r load files}
germline <- read_tsv(file.path(root_dir, params$germ_in)) %>%
  bind_rows(read_tsv(file.path(data_dir,
                               "pbta-merged-plp-variants-autogvp-abridged-lowVAF-cpgs.tsv"))) %>%
  dplyr::rename(Matched_Norm_Sample_Barcode = Kids_First_Biospecimen_ID_normal,
                Hugo_Symbol = gene_symbol_vep,
                HGVSc_germ = HGVSc,
                HGVSp_germ = HGVSp) 

integrated_hist <- read_tsv(histologies_file)
opc_hist <- read_tsv(opc_hist_file, guess_max = 3000) %>%
  filter(!is.na(pathology_diagnosis),
         experimental_strategy == "WGS",
         tumor_descriptor != "Derived Cell Line")

# CPGs
cpg <- read_lines(cpg_file)

# load maf
maf <- data.table::fread(file.path(root_dir, params$maf_in), data.table = F) %>%
  filter(Tumor_Sample_Barcode %in% opc_hist$Kids_First_Biospecimen_ID)

# CPGs without somatic hits
setdiff(cpg, unique(maf$Hugo_Symbol))

# table of gene counts by oncogenic effect
maf %>%
  count(ONCOGENIC) %>%
  arrange(-n)

# extract oncogenic/likely oncogenic and/or other LOF variants
lof_hits <- maf %>%
  filter(ONCOGENIC %in% c("Likely Oncogenic", "Oncogenic", "Resistance") |
           Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins",
                                         "Missense_Mutation", "Nonsense_Mutation",
                                         "Splice_Site", "Translation_Start_Site")) %>%
  write_tsv(file.path(results_dir, "pbta-oncokb-oncogenic-lof-maf.tsv"))

# prevalence by gene
lof_hits %>%
  count(Hugo_Symbol) %>%
  arrange(-n)

# prevalence by variant in oncokb
lof_hits %>%
  count(VARIANT_IN_ONCOKB) %>%
  arrange(-n)

# prevalence by variant in oncokb
lof_hits %>%
  count(MUTATION_EFFECT) %>%
  arrange(-n)

# how many are hotspots?
lof_hits %>%
  count(HotSpotAllele) %>%
  arrange(-n)

```

### Integrate with germline data

```{r two hits same gene}
# collapse calls a bit
lof_hits_subset <- lof_hits %>%
  mutate(VAF = t_alt_count / t_depth) %>%
  select(Matched_Norm_Sample_Barcode, Tumor_Sample_Barcode, Hugo_Symbol, Transcript_ID, HGVSc, HGVSp, VAF, SIFT, PolyPhen, Variant_Classification, HotSpotAllele, ONCOGENIC, MUTATION_EFFECT, VARIANT_IN_ONCOKB)

# two hit CPG matches
germline_somatic_gene_matches <- germline %>%
  inner_join(lof_hits_subset, by = c("Matched_Norm_Sample_Barcode", "Hugo_Symbol")) %>%
  write_tsv(file.path(results_dir, "germline-somatic-two-gene-hits.tsv"))
```

### List germline + somatic gene hits by sample

```{r gene hits by patient}
collapsed <- germline %>%
  filter(Hugo_Symbol %in% cpg) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_normal = Matched_Norm_Sample_Barcode) %>%
  group_by(Kids_First_Biospecimen_ID_normal) %>%
  summarise(Hugo_Symbol = str_c(sort(unique(Hugo_Symbol)), collapse = "; ")) %>%
  left_join(integrated_hist)

collapsed_lof <- lof_hits_subset %>%
  dplyr::rename(Kids_First_Biospecimen_ID_normal = Matched_Norm_Sample_Barcode,
                Kids_First_Biospecimen_ID_tumor_maf = Tumor_Sample_Barcode,
                Hugo_Symbol_somatic = Hugo_Symbol) %>%
  group_by(Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor_maf) %>%
  summarise(Hugo_Symbol_somatic = str_c(sort(unique(Hugo_Symbol_somatic)), collapse = "; "))

collapsed_germ_som <- collapsed %>%
  left_join(collapsed_lof, by = "Kids_First_Biospecimen_ID_normal") %>%
  write_tsv(file.path(results_dir, "germline-somatic-collapsed-by-gene.tsv"))
```


```{r}
germline_sv <- read_tsv(plp_sv_file) %>%
  dplyr::rename(Matched_Norm_Sample_Barcode = Kids_First_Biospecimen_ID_normal,
                Hugo_Symbol = Hugo_Symbol_cpg) 

# two hit CPG matches
germline_sv_somatic_gene_matches <- germline_sv %>%
  inner_join(onco_hits_subset, by = c("Matched_Norm_Sample_Barcode", "Hugo_Symbol")) %>%
  write_tsv(file.path(results_dir, "germline-sv-somatic-two-gene-hits.tsv"))
```

Create lollipop plots of germline variants + somatic mutations. First, modify germline variant df to be compatible with maf columns:

```{r}
germ_maf <- germline_somatic_gene_matches %>%
  dplyr::select(-HGVSc, -HGVSp, -Variant_Classification) %>%
  # rename columns to match maf columns
  dplyr::rename(Chromosome = chr,
                Start_Position = start,
                Variant_Classification = variant_classification_vep,
                Reference_Allele = ref,
                HGVSc = HGVSc_germ,
                HGVSp_Short = HGVSp_germ) %>%
  # modify variant classifications to match those in maf
  dplyr::mutate(Variant_Classification = case_when(
    Variant_Classification == "missense_variant" ~ "Missense_Mutation",
    grepl("stop_gained", Variant_Classification) ~ "Nonsense_Mutation",
    grepl("splice_acceptor|splice_donor", Variant_Classification) ~ "Splice_Site",
    Variant_Classification == "frameshift_variant" & nchar(alt) > nchar(Reference_Allele) ~ "Frame_Shift_Ins",
    Variant_Classification == "frameshift_variant" & nchar(alt) < nchar(Reference_Allele) ~ "Frame_Shift_Del",
    Variant_Classification == "inframe_deletion" ~ "In_Frame_Del",
    TRUE ~ Variant_Classification
  )) %>%
  # add variant_type column
  dplyr::mutate(Variant_Type = case_when(
    nchar(alt) == nchar(Reference_Allele) ~ "SNP",
    nchar(alt) < nchar(Reference_Allele) ~ "DEL",
    nchar(alt) > nchar(Reference_Allele) ~ "INS"
  )) %>%
  # Replace AA symbols with one-letter symbols
  dplyr::mutate(HGVSp_Short = str_replace(HGVSp_Short, "Leu", "L"),
                HGVSp_Short = str_replace(HGVSp_Short, "Val", "V"),
                HGVSp_Short = str_replace(HGVSp_Short, "Arg", "R"),
                HGVSp_Short = str_replace(HGVSp_Short, "Ser", "S"),
                HGVSp_Short = str_replace(HGVSp_Short, "Asn", "N"),
                HGVSp_Short = str_replace(HGVSp_Short, "Met", "M"),
                HGVSp_Short = str_replace(HGVSp_Short, "Lys", "K"),
                HGVSp_Short = str_replace(HGVSp_Short, "Ter", "*"),
                HGVSp_Short = str_replace(HGVSp_Short, "Gln", "G"),
                HGVSp_Short = str_replace(HGVSp_Short, "Tyr", "Y"),
                HGVSp_Short = str_replace(HGVSp_Short, "Phe", "F"),
                HGVSp_Short = str_replace(HGVSp_Short, "Ala", "A")) %>%
  dplyr::select(Matched_Norm_Sample_Barcode, Hugo_Symbol,
                Tumor_Sample_Barcode,
                Chromosome, Start_Position, Variant_Classification,
                Reference_Allele, HGVSc, HGVSp_Short, Variant_Type)

merged_maf <- maf %>%
  bind_rows(germ_maf)

write_tsv(merged_maf, file.path(results_dir, "germ-somatic-merged-maf.tsv"))
```

Re-load merged_maf as maf object and generate lollipop plots

```{r}
merged_maf <- maftools::read.maf(file.path(results_dir, "germ-somatic-merged-maf.tsv"))

ids <- germline_somatic_gene_matches %>%
  distinct(Tumor_Sample_Barcode) %>%
  pull(Tumor_Sample_Barcode)

for (id in ids){
  
  gene <- germline_somatic_gene_matches %>%
    dplyr::filter(Tumor_Sample_Barcode == id) %>%
    pull(Hugo_Symbol) %>%
    unique()
  
  to_rm <- unique(maf$Tumor_Sample_Barcode[maf$Tumor_Sample_Barcode != id])

  plot_maf <- maftools::filterMaf(merged_maf, tsb = to_rm)
  
  pdf(file.path(plot_dir,
                   glue::glue("{id}-{gene}-lollipop.pdf")),
         width = 6, height = 4)
  
  lollipopPlot(plot_maf, gene = gene, AACol = "HGVSp_Short", 
               showDomainLabel = FALSE, labPosAngle = 55, 
               labelPos = "all", legendTxtSize = 1, 
               labPosSize = 1.25,
               pointSize = 2,
               showMutationRate = FALSE,
               collapsePosLabel = TRUE)
  
  dev.off()
  
}
```

Print session info
```{r}
sessionInfo()
```