---
title: "Assess somatic CNV and LOH second hits"
output: 
  html_notebook:
  toc: TRUE
toc_float: TRUE
author: Ryan Corbett
---

This analysis identifies patients with germline P-LP variants in cancer predisposition genes that show subsequent seconds CNV or loss of heterozygosity (LOH) hits in tumors 

Set libraries and file paths
```{r load libraries}
# load libraries
library(tidyverse)
library(ggplot2)
library(ggpubr)

# define directories 
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "two-hits")
results_dir <- file.path(analysis_dir, "results")
plot_dir <- file.path(analysis_dir, "plots")
data_dir <- file.path(root_dir, "data")

# If the input and results directories do not exist, create it
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

source(file.path(root_dir, "figures", "theme.R"))
source(file.path(analysis_dir, "util", "cnv_loh_functions.R"))
```

Define file paths
```{r set file paths}
# germline P-LP calls 
plp_file <- file.path(data_dir, 
                      "pbta-merged-plp-variants-autogvp-abridged.tsv")

plp_lowVAF_file <- file.path(data_dir,
                             "pbta-merged-plp-variants-autogvp-abridged-lowVAF-cpgs.tsv")

plp_sv_file <- file.path(data_dir,
                         "pbta_germline_svs.tsv")

# add histologies file 
histologies_file <- file.path(root_dir, "analyses", 
                                  "collapse-tumor-histologies", "results", 
                                  "germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv")

# cnv file with annotation 
cnv_auto_file <- file.path(data_dir, "consensus_seg_annotated_cn_autosomes.tsv.gz")
cnv_xy_file <- file.path(data_dir, "consensus_seg_annotated_cn_x_and_y.tsv.gz")

# LOH variant and gene files
loh_variant_file <- file.path(data_dir, "pbta-plp-all-loh.tsv")
loh_gene_file <- file.path(data_dir, "pbta-all-gene-loh.tsv.gz")

cnvkit_file <- file.path(data_dir, "cnv-cnvkit-cns.tsv.gz")

# cpg file
cpg_file <- file.path(root_dir, "analyses", "oncokb-annotation", "input", "cpg.txt")
hgat_file <- file.path(root_dir, "analyses", "variant-distribution", "input", "hgat_goi_list.tsv")
lgat_file <- file.path(root_dir, "analyses", "variant-distribution", "input", "lgat_goi_list.tsv")
embryonal_file <- file.path(root_dir, "analyses", "variant-distribution", "input", "embryonal-tumor_goi_list.tsv")
other_file <- file.path(root_dir, "analyses", "variant-distribution", "input", "other_goi_list.tsv")
```

Read in PLP, histology, and cpg files. Subset P-LP calls to CPGs only and add histology
```{r get all matched tumors}
hist <- read_tsv(histologies_file)

# Load abridged P-LP file and append variant coordinates
germline <- read_tsv(plp_file) %>%
  bind_rows(read_tsv(plp_lowVAF_file)) %>%
  dplyr::rename(Hugo_Symbol = gene_symbol_vep) %>%
  filter(Kids_First_Biospecimen_ID_normal %in% hist$Kids_First_Biospecimen_ID_normal)

germline_sv <- read_tsv(plp_sv_file) %>%
  dplyr::rename(Hugo_Symbol = Hugo_Symbol_cpg) %>%
  filter(Kids_First_Biospecimen_ID_normal %in% hist$Kids_First_Biospecimen_ID_normal)

# CPGs
cpgs <- read_lines(cpg_file)
hgat_drivers <- read_lines(hgat_file)
lgat_drivers <- read_lines(lgat_file)
embryonal_drivers <- read_lines(embryonal_file)
other_drivers <- read_lines(other_file)
somatic_drivers <- c(hgat_drivers, lgat_drivers, embryonal_drivers, other_drivers)

# filter PLP calls for cpgs, and add plot_group and molecular_subtype
germline <- germline %>%
  filter(Hugo_Symbol %in% c(cpgs, somatic_drivers)) %>%
  left_join(hist[,c("Kids_First_Participant_ID", "Kids_First_Biospecimen_ID_normal",
                          "Kids_First_Biospecimen_ID_tumor",
                          "plot_group", "molecular_subtype")], 
            by = "Kids_First_Biospecimen_ID_normal") %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, Kids_First_Participant_ID, 
                Kids_First_Biospecimen_ID_tumor, 
                plot_group, molecular_subtype,
                Hugo_Symbol, autogvp_call, autogvp_call_reason,
                chr, start, ref, alt, HGVSc, HGVSp) %>%
  distinct(Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor, Hugo_Symbol, start, .keep_all = T)

germline_sv <- germline_sv %>%
  left_join(hist[,c("Kids_First_Participant_ID", "Kids_First_Biospecimen_ID_normal",
                          "Kids_First_Biospecimen_ID_tumor",
                          "plot_group", "molecular_subtype")], 
            by = "Kids_First_Biospecimen_ID_normal") %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, Kids_First_Participant_ID, 
                Kids_First_Biospecimen_ID_tumor, 
                plot_group, molecular_subtype,
                Hugo_Symbol, Chromosome, Start, End, Length,
                Type, Classification) %>%
  distinct(Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor, Hugo_Symbol, .keep_all = T)
```

Load CNV files, and identify patients with germline PLP variant and somatic CNV in same CPG
```{r get cnv data}
# filter cnv to patients in cohort and CPG calls only
cnv_auto <- read_tsv(cnv_auto_file) %>%
  dplyr::filter(biospecimen_id %in% germline$Kids_First_Biospecimen_ID_tumor & gene_symbol %in% c(cpgs, somatic_drivers)) %>%
  dplyr::rename("Hugo_Symbol" = gene_symbol,
         "Kids_First_Biospecimen_ID_tumor" = biospecimen_id,
         "CNV_status_somatic" = status,
         "copy_number_somatic" = copy_number,
         "ploidy_somatic" = "ploidy") %>%
  dplyr::select(Hugo_Symbol, Kids_First_Biospecimen_ID_tumor,
         CNV_status_somatic)

cnv_xy <- read_tsv(cnv_xy_file) %>%
  dplyr::filter(biospecimen_id %in% germline$Kids_First_Biospecimen_ID_tumor & gene_symbol %in% c(cpgs, somatic_drivers)) %>%
  dplyr::rename("Hugo_Symbol" = gene_symbol,
         "Kids_First_Biospecimen_ID_tumor" = biospecimen_id,
         "CNV_status_somatic" = status,
         "copy_number_somatic" = copy_number,
         "ploidy_somatic" = "ploidy") %>%
  dplyr::select(Hugo_Symbol, Kids_First_Biospecimen_ID_tumor,
         CNV_status_somatic)

# merge cnv tables
cnv <- cnv_auto %>%
  bind_rows(cnv_xy) %>%
  dplyr::filter(Kids_First_Biospecimen_ID_tumor %in% hist$Kids_First_Biospecimen_ID_tumor)

# merge germline SNV and somatic CNV calls by bs and gene id
germline_somatic_cnvs <- germline %>%
  left_join(cnv, by = c("Kids_First_Biospecimen_ID_tumor", "Hugo_Symbol")) %>%
  dplyr::mutate(CNV_status_somatic = case_when(
    is.na(CNV_status_somatic) ~ "neutral",
    TRUE ~ CNV_status_somatic
  )) %>%
  distinct()

germline_sv_somatic_cnvs <- germline_sv %>%
  left_join(cnv, by = c("Kids_First_Biospecimen_ID_tumor", "Hugo_Symbol")) %>%
  dplyr::mutate(CNV_status_somatic = case_when(
    is.na(CNV_status_somatic) ~ "neutral",
    TRUE ~ CNV_status_somatic
  )) %>%
  distinct()
```

Read in LOH results, and append LOH scores, odds ratios, and p-values to rows of `germline_somatic_cnvs`
```{r get variant loh data}
loh <- read_tsv(loh_variant_file) %>%
  dplyr::filter(start %in% germline$start) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_normal = BS_ID,
                Kids_First_Biospecimen_ID_tumor = BS_ID_tumor,
                Hugo_Symbol = gene) %>%
  dplyr::filter(!is.na(proband_tumor_alt_depth),
                Kids_First_Biospecimen_ID_tumor %in% hist$Kids_First_Biospecimen_ID_tumor) %>%
  dplyr::mutate(loh_score = proband_tumor_vaf - proband_germline_vaf,
                loh_OR = 0,
                loh_pvalue = 0) %>%
  distinct(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, start, Kids_First_Biospecimen_ID_tumor,
           .keep_all = TRUE)

# Calculate LOH OR and corresponding p-values
for (i in 1:nrow(loh)){
  
  germline_ref_count <- loh$proband_germline_ref_depth[i]
  germline_alt_count <- loh$proband_germline_alt_depth[i]
  tumor_ref_count <- loh$proband_tumor_depth[i] - loh$proband_tumor_alt_depth[i]
  tumor_alt_count <- loh$proband_tumor_alt_depth[i]
  
  mat <- matrix(c(germline_ref_count, germline_alt_count,
                  tumor_ref_count, tumor_alt_count),
                2,2)
  
  loh$loh_OR[i] <- fisher.test(mat)$estimate
  loh$loh_pvalue[i] <- fisher.test(mat)$p.value
  
}
```


```{r}
cnvkit <- read_tsv(cnvkit_file) %>%
  dplyr::filter(ID %in% hist$Kids_First_Biospecimen_ID_tumor)

loh <- add_cnvkit(loh, cnvkit)

loh <- loh %>%
  dplyr::mutate(loh_status = case_when(
    (proband_tumor_vaf > 0.75 & loh_pvalue < 0.05) | 
      (proband_tumor_vaf > 0.25 & cnvkit_cn2 == 0 & cnvkit_cn1 != 0) ~ "LOH",
    TRUE ~ "No LOH"
  )) %>%
  arrange(Kids_First_Biospecimen_ID_normal) %>%
  write_tsv(file.path(results_dir,
                "pbta-loh-sig-calculated.tsv"))
```

Plot tumor VAF vs germline VAF
```{r}
loh %>%
  ggplot(aes(x = proband_germline_vaf, y = proband_tumor_vaf,
                colour = loh_status)) +
  geom_point() +
  labs(x = "Germline VAF",
       y = "Tumor VAF", color = "LOH status") +
  scale_color_manual(values = c("LOH" = "orangered", "No LOH" = "gray")) + 
  theme_Publication()

ggsave(file.path(plot_dir, "tumor-vaf-vs-germline-vaf.pdf"),
       width = 5, height = 4, units = "in")
```

Append somatic LOH results for germline variants
```{r}
germline_somatic_cnvs_loh <- germline_somatic_cnvs %>%
  left_join(loh[,c("Kids_First_Biospecimen_ID_tumor", 
                   "chr", "start", "proband_germline_vaf", "proband_tumor_vaf", 
                   "loh_score", "loh_OR", "loh_pvalue", "cnvkit_log2", "cnvkit_cn",
                   "cnvkit_cn1", "cnvkit_cn2")], 
            by = c("Kids_First_Biospecimen_ID_tumor", "chr", "start")) %>%
  distinct()
```

Load gene LOH data and append to `germline_somatic_cnvs_loh`
```{r get gene loh data}
loh_gene <- read_tsv(loh_gene_file) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_normal = BS_ID,
                Kids_First_Biospecimen_ID_tumor = BS_ID_tumor,
                Hugo_Symbol = gene,
                gene_LOH = mean_LOH_score) %>%
  dplyr::filter(Hugo_Symbol %in% c(cpgs, somatic_drivers)) %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor, Hugo_Symbol,
         gene_LOH) %>%
  dplyr::filter(Kids_First_Biospecimen_ID_tumor %in% hist$Kids_First_Biospecimen_ID_tumor)

germline_somatic_cnvs_loh_gene <- germline_somatic_cnvs_loh %>%
  left_join(loh_gene)

germline_sv_somatic_cnvs_loh_gene <- germline_sv_somatic_cnvs %>%
  left_join(loh_gene)

germline_sv_somatic_cnvs_loh_gene <- add_cnvkit(germline_sv_somatic_cnvs_loh_gene,
                                                cnvkit)
```

Save CNV and LOH scores for all samples to output
```{r save results}
germline_somatic_cnvs_loh_gene %>%
  dplyr::filter(Hugo_Symbol %in% cpgs) %>%
  arrange(Kids_First_Biospecimen_ID_normal) %>%
  write_tsv(file.path(results_dir, "germline-somatic-cnv-loh-cpgs.tsv"))

germline_somatic_cnvs_loh_gene %>%
  dplyr::filter(!Hugo_Symbol %in% cpgs) %>%
  arrange(Kids_First_Biospecimen_ID_normal) %>%
  write_tsv(file.path(results_dir, "germline-somatic-cnv-loh-somatic-non-cpgs.tsv"))

germline_sv_somatic_cnvs_loh_gene %>%
  arrange(Kids_First_Biospecimen_ID_normal) %>%
  write_tsv(file.path(results_dir, "germline-sv-somatic-cnv-loh-somatic-cpgs.tsv"))
```

Create df to plot LOH by CNV status
```{r}
cnv_loh_plot_df <- germline_somatic_cnvs_loh_gene %>%
  dplyr::filter(Kids_First_Biospecimen_ID_tumor %in% hist$Kids_First_Biospecimen_ID_tumor,
                Hugo_Symbol %in% cpgs) %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, 
                CNV_status_somatic, loh_score, gene_LOH) %>%
  bind_rows(germline_sv_somatic_cnvs_loh_gene %>%
              dplyr::select(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, 
                CNV_status_somatic, gene_LOH)) %>%
  distinct(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, .keep_all = TRUE) %>%
  dplyr::mutate(CNV_status_somatic = fct_relevel(CNV_status_somatic,
                                                 c("gain", "neutral", "loss")))
```

Plot variant LOH by CNV status
```{r}
cnv_loh_plot_df %>%
  ggplot(aes(x = CNV_status_somatic, y = abs(loh_score),
              fill = CNV_status_somatic)) +
  geom_jitter(shape = 21, width = 0.2, size = 2.5, alpha = 0.85,
              show.legend = FALSE) +
  geom_boxplot(alpha = 0.05, outlier.shape = NA,
               show.legend = FALSE) +
  stat_compare_means(method = "wilcox",
                     comparisons = list(c("gain", "loss"),
                                        c("loss", "neutral"),
                                        c("gain", "neutral")),
                     method.args = list(alternative = "greater")) + 
  geom_hline(yintercept = 0.2, linetype = "dashed", color = "gray") + 
  labs(x = "Germline P/LP variant", y = "abs(variant LOH score)", color = "Histology") +
  ylim(c(0,1)) +
  theme_Publication()

  ggsave(file.path(plot_dir, "variant-LOH-by-cnv-status.pdf"),
                 width = 6, height = 4)
```  

Plot gene LOH by CNV status
```{r}
cnv_loh_plot_df %>%
  ggplot(aes(x = CNV_status_somatic, y = abs(gene_LOH),
              fill = CNV_status_somatic)) +
  geom_jitter(shape = 21, width = 0.2, size = 2.5, alpha = 0.85,
              show.legend = FALSE) +
  geom_boxplot(alpha = 0.05, outlier.shape = NA,
               show.legend = FALSE) +
  stat_compare_means(method = "wilcox",
                     comparisons = list(c("gain", "loss"),
                                        c("loss", "neutral"),
                                        c("gain", "neutral")),
                     method.args = list(alternative = "greater")) + 
  geom_hline(yintercept = 0.2, linetype = "dashed", color = "gray") + 
  labs(x = "Germline P/LP variant", y = "gene LOH score", color = "Histology") +
  ylim(c(0,0.75)) +
  theme_Publication()

  ggsave(file.path(plot_dir, "gene-LOH-by-cnv-status.pdf"),
                 width = 6, height = 4)
```

## Plotting LOH by gene and P-LP status

For plotting, we will use LOH scores from one matched tumor sample per patient, specified in the cohort histologies file
```{r subset for plotting}
germline_all <- germline %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, plot_group) %>%
  bind_rows(germline_sv %>% dplyr::select(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, plot_group))

loh_gene <- loh_gene %>%
  # filter for only tumor IDs in cohort histologies
  dplyr::filter(Kids_First_Biospecimen_ID_tumor %in% hist$Kids_First_Biospecimen_ID_tumor) %>%
  # add column indicating germline P/LP status for each gene and patient
  dplyr::mutate(is_plp = case_when(
    glue::glue("{Kids_First_Biospecimen_ID_normal}:{Hugo_Symbol}") %in% glue::glue("{germline_all$Kids_First_Biospecimen_ID_normal}:{germline_all$Hugo_Symbol}") ~ "Yes",
    TRUE ~ "No"
  )) %>%
  # append plot_group and hex code for plotting
  left_join(germline_all[,c("Kids_First_Biospecimen_ID_normal", "plot_group")]) %>%
  left_join(hist[,c("plot_group", "plot_group_hex")]) %>%
  distinct(Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor, Hugo_Symbol, .keep_all = TRUE)
```

Look at LOH score differences in P-LP carriers vs non-carriers within histologies 
```{r plot LOH by P/LP status within histologies}
# define group_gene column
loh_gene <- loh_gene %>%
  dplyr::mutate(group_gene = glue::glue("{Hugo_Symbol}: {plot_group}"))

# define color palette
plot_group_palette <- loh_gene$plot_group_hex
names(plot_group_palette) <- loh_gene$plot_group

# Identify histology-gene pairs for which there are at least 3 patients with P/LP variants in group
plot_group_gene <- loh_gene %>%
  distinct(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, .keep_all = TRUE) %>%
  filter(is_plp == "Yes") %>%
  count(group_gene) %>%
  dplyr::filter(n >= 2) %>%
  pull(group_gene)

compare_means_result <- loh_gene %>%
  dplyr::filter(group_gene %in% plot_group_gene & group_gene != "NF1: Neurofibroma plexiform") %>%
  group_by(group_gene) %>%
  summarize(p_value = wilcox.test(gene_LOH ~ is_plp, alternative = "less")$p.value)

to_plot <- compare_means_result %>%
  dplyr::filter(p_value < 0.05) %>%
  pull(group_gene)

# plot LOH scores by P/LP status within histologies for selected gene-histology pairs
loh_gene %>%
  dplyr::filter(group_gene %in% to_plot) %>%
  ggplot(aes(x = is_plp, y = gene_LOH)) + 
  geom_jitter(aes(fill = plot_group), shape = 21, 
              width = 0.2, size = 2.5, alpha = 0.85,
              show.legend = FALSE) +
  geom_boxplot(alpha = 0.05, outlier.shape = NA,
               show.legend = FALSE) +
  stat_compare_means(method = "wilcox",
                     comparisons = list(c("Yes", "No")),
                     method.args = list(alternative = "greater")) + 
  labs(x = "Germline P/LP variant", y = "gene LOH score",
       fill = "Histology") +
  ylim(c(0,0.75)) +
  scale_fill_manual(values = plot_group_palette, breaks = sort(names(plot_group_palette)), na.translate = F) +
  facet_wrap(~group_gene, nrow = 2,
             labeller = labeller(group_gene = label_wrap_gen(18))) + 
  theme_Publication()
  
  ggsave(file.path(plot_dir, "hist-gene-LOH-plp-vs-noplp.pdf"),
                   width = 9, height = 6)
```

```{r}
sessionInfo()
```