---
title: "Calculate somatic hit enrichment among cancer predisposition genes"
output: 
  html_notebook:
  toc: TRUE
toc_float: TRUE
author: Ryan Corbett
---

This script tests for enrichment of somatic alterations in cancer predisposition genes among patients with germline P-LP variants in the same gene, relative to the rest of the cohort

# Load libraries
```{r load libraries, set up directories}
suppressPackageStartupMessages({
  library(ggplot2)
  library(tidyverse)
  library(ComplexHeatmap)
  library(circlize)
  library(ggpubr)
})

# Set up directories
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "two-hits")
plot_dir <- file.path(analysis_dir, "plots")
results_dir <- file.path(analysis_dir, "results")

# If the directory does not exist, create it
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

if (!dir.exists(plot_dir)) {
  dir.create(plot_dir, recursive = TRUE)
}

# source publication theme
source(file.path(root_dir, "figures", "theme.R"))
```

Set input file paths
```{r}
plp_file <- file.path(data_dir, 
                      "pbta_germline_plp_calls_autogvp_abridged_11062023.tsv")

hist_file <- file.path(data_dir, "histologies.tsv")

maf_file <- file.path(root_dir, "analyses", "oncokb-annotation", "results", "snv-consensus-plus-hotspots-goi-oncokb.maf.tsv")
cnv_file <- file.path(data_dir, "consensus_seg_annotated_cn_autosomes.tsv.gz")
loh_file <- file.path(data_dir, "pbta-all-gene-loh.tsv.gz")
expr_file <- file.path(data_dir, "gene-expression-rsem-tpm-collapsed.rds")

specimens_file <- file.path(data_dir, "independent-specimens.rnaseqpanel.primary-plus.tsv")
```

# Load files
```{r load figure theme, files}
# read in histology file
hist_cohort <- read_tsv(file.path(root_dir, "analyses", 
                                  "collapse-tumor-histologies", "results",
                           "germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv"), 
                        show_col_types = FALSE)

hist_all <- read_tsv(hist_file, guess_max = 1000000)
  
hist_all_genomic <- hist_all %>%
  dplyr::filter(sample_type == "Tumor" & experimental_strategy %in% c("WGS", "WXS"))

hist_all_rna <- hist_all %>%
  dplyr::filter(sample_type == "Tumor" & experimental_strategy == "RNA-Seq")

tumor_normal_genome_df <- hist_cohort %>%
  dplyr::select(-Kids_First_Biospecimen_ID_tumor) %>%
  left_join(hist_all_genomic[,c("Kids_First_Participant_ID", "Kids_First_Biospecimen_ID")]) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_tumor = Kids_First_Biospecimen_ID) %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_tumor)

tumor_normal_rna_df <- hist_cohort %>%
  left_join(hist_all_rna[,c("Kids_First_Participant_ID", "Kids_First_Biospecimen_ID")]) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_rna = Kids_First_Biospecimen_ID) %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, Kids_First_Biospecimen_ID_rna)

maf <- read_tsv(maf_file)

plp <- read_tsv(plp_file) %>%
  dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist_cohort$Kids_First_Biospecimen_ID_normal) %>%
  dplyr::rename(Hugo_Symbol = gene_symbol_vep)

cnv <- read_tsv(cnv_file) %>%
  dplyr::filter(biospecimen_id %in% tumor_normal_genome_df$Kids_First_Biospecimen_ID_tumor) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_tumor = biospecimen_id) %>%
  left_join(tumor_normal_genome_df) %>%
  distinct() %>%
  dplyr::filter(gene_symbol %in% plp$Hugo_Symbol)
  
loh <- read_tsv(loh_file) %>%
  dplyr::filter(Kids_First_Biospecimen_ID_normal %in% tumor_normal_genome_df$Kids_First_Biospecimen_ID_normal) %>%
  dplyr::filter(Hugo_Symbol %in% plp$Hugo_Symbol)

expr <- readRDS(expr_file)
# subset `expr` to only include BS IDs in hist
subset_expr <- expr[,intersect(tumor_normal_rna_df$Kids_First_Biospecimen_ID_rna, colnames(expr))]
```

calculate z-scores from TPM data
```{r}
# log2(x + 1) transform the expression matrix
log_expression <- log2(subset_expr + 1)

# Scale the gene values -- scale() works on the columns, hence the transpose
z_scored_expression <- scale(t(log_expression),
                             center = TRUE,
                             scale = TRUE)
z_scored_expression <- as.data.frame(z_scored_expression)

z_scored_expression <- z_scored_expression %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Hugo_Symbol") %>%
  gather(key = "Kids_First_Biospecimen_ID", value = "expr_zscore", -Hugo_Symbol) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_rna = Kids_First_Biospecimen_ID) %>%
  left_join(tumor_normal_rna_df) %>%
  dplyr::filter(Hugo_Symbol %in% plp$Hugo_Symbol)

```

Identify cancer predisposition genes with P-LP variants in >= 4 patients
```{r}
cpg <- read_lines(file.path(root_dir, "analyses", "oncokb-annotation", "input", "cpg.txt"))

genes <- plp %>%
  distinct(Kids_First_Biospecimen_ID_normal, Hugo_Symbol, .keep_all = T) %>%
  dplyr::count(Hugo_Symbol) %>%
  dplyr::filter(n >= 4 & Hugo_Symbol %in% cpg) %>%
  pull(Hugo_Symbol)

```

Create empty data frames to store enrichment stats
```{r}
empty_df <- data.frame(row.names = genes,
                     alteration = rep(" ", length(genes)),
                     n_plp_alt = rep(0, length(genes)), 
                     n_plp_no_alt = rep(0, length(genes)),
                     n_no_plp_alt = rep(0, length(genes)),
                     n_no_plp_no_alt = rep(0, length(genes)),
                     oddsRatio = rep(0, length(genes)),
                     p = rep(0, length(genes)),
                     CI_lower = rep(0, length(genes)),
                     CI_upper = rep(0, length(genes)),
                     adjusted_p = rep(0, length(genes)))

snv_df <- empty_df %>%
  dplyr::mutate(alteration = "Oncogenic SNV")

cnv_loss_df <- empty_df %>%
  mutate(alteration = "CN loss")

cnv_gain_df <- empty_df %>%
  dplyr::mutate(alteration = "CN gain")

loh_df <- empty_df %>%
  dplyr::mutate(alteration = "LOH > 0.25")

expr_gain_df <- empty_df %>%
  dplyr::mutate(alteration = "TPM z-score > 2")

expr_loss_df <- empty_df %>%
  dplyr::mutate(alteration = "TPM z-score < -2")

```

Loop through genes and calculate enrichment of somatic alterations among patients with germline P-LP variant in same gene
```{r}
for (gene in genes){
  
  # identify samples with P-LP variant in gene
  plp_gene_bs <- plp %>%
    dplyr::filter(Hugo_Symbol == gene) %>%
    pull(Kids_First_Biospecimen_ID_normal) 

  # Filter MAF for only variants in gene
  gene_maf <- maf %>%
      dplyr::filter(Hugo_Symbol == gene & ONCOGENIC %in% c("Oncogenic", "Likely Oncogenic")) 
  
  # create contingency table of plp T/F and somatic alteration T/F
  snv_mat <- table(is_plp = hist_cohort$Kids_First_Biospecimen_ID_normal %in% plp_gene_bs, 
                   is_alt = factor(hist_cohort$Kids_First_Biospecimen_ID_normal %in% gene_maf$Matched_Norm_Sample_Barcode, levels = c("FALSE", "TRUE")))
  
  # add stats to df 
  snv_df[gene, c("n_no_plp_no_alt", "n_plp_no_alt", "n_no_plp_alt", "n_plp_alt")] = unlist(snv_mat)
  snv_df[gene, c("oddsRatio", "p", "CI_lower", "CI_upper")] <- unlist(fisher.test(snv_mat)[c("estimate", "p.value", "conf.int")])
  
  # filter cnv df for samples with loss in gene
  gene_cnv_loss <- cnv %>%
    dplyr::filter(gene_symbol == gene & status == "loss")
  
  # create P/LP and somatic alteration contingency table and store stats in df
  cnv_loss_mat <- table(is_plp = hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% cnv$Kids_First_Biospecimen_ID_normal] %in% plp_gene_bs,
                        is_alt = factor(hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% cnv$Kids_First_Biospecimen_ID_normal] %in% gene_cnv_loss$Kids_First_Biospecimen_ID_normal, levels = c("FALSE", "TRUE")))
  
  cnv_loss_df[gene, c("n_no_plp_no_alt", "n_plp_no_alt", "n_no_plp_alt", "n_plp_alt")] = unlist(cnv_loss_mat)
  cnv_loss_df[gene, c("oddsRatio", "p", "CI_lower", "CI_upper")] <- unlist(fisher.test(cnv_loss_mat)[c("estimate", "p.value", "conf.int")])
  
  # Repeat for copy number gains
  gene_cnv_gain <- cnv %>%
    dplyr::filter(gene_symbol == gene & status == "gain")
  
  cnv_gain_mat <- table(is_plp = hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% cnv$Kids_First_Biospecimen_ID_normal] %in% plp_gene_bs,
                        is_alt = factor(hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% cnv$Kids_First_Biospecimen_ID_normal] %in% gene_cnv_gain$Kids_First_Biospecimen_ID_normal, levels = c("FALSE", "TRUE")))
  
  cnv_gain_df[gene, c("n_no_plp_no_alt", "n_plp_no_alt", "n_no_plp_alt", "n_plp_alt")] = unlist(cnv_gain_mat)
  cnv_gain_df[gene, c("oddsRatio", "p", "CI_lower", "CI_upper")] <- unlist(fisher.test(cnv_gain_mat)[c("estimate", "p.value", "conf.int")])
  
  # Filter LOH df for samples exhibiting LOH in gene
  gene_loh <- loh %>%
    dplyr::filter(Hugo_Symbol == gene & LOH_score > 0.25)
  
  # Create contingency table and store stats for LOH enrichment 
  loh_mat <- table(is_plp = hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% loh$Kids_First_Biospecimen_ID_normal] %in% plp_gene_bs,
                        is_alt = factor(hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% loh$Kids_First_Biospecimen_ID_normal] %in% gene_loh$Kids_First_Biospecimen_ID_normal, levels = c("FALSE", "TRUE")))
  
  loh_df[gene, c("n_no_plp_no_alt", "n_plp_no_alt", "n_no_plp_alt", "n_plp_alt")] = unlist(loh_mat)
  loh_df[gene, c("oddsRatio", "p", "CI_lower", "CI_upper")] <- unlist(fisher.test(loh_mat)[c("estimate", "p.value", "conf.int")])

  # Filter expr data for samples with expression loss in gene
  gene_expr_loss <- z_scored_expression %>%
    dplyr::filter(Hugo_Symbol == gene & expr_zscore < -2)
  
  # create P/LP and expr loss contingency table and store stats in expr_loss_df
  expr_loss_mat <- table(is_plp = hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% z_scored_expression$Kids_First_Biospecimen_ID_normal] %in% plp_gene_bs,
                   is_alt = factor(hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% z_scored_expression$Kids_First_Biospecimen_ID_normal] %in% gene_expr_loss$Kids_First_Biospecimen_ID_normal, levels = c("FALSE", "TRUE")))
  
  expr_loss_df[gene, c("n_no_plp_no_alt", "n_plp_no_alt", "n_no_plp_alt", "n_plp_alt")] = unlist(expr_loss_mat)
  expr_loss_df[gene, c("oddsRatio", "p", "CI_lower", "CI_upper")] <- unlist(fisher.test(expr_loss_mat)[c("estimate", "p.value", "conf.int")])

  # Repeat for expression gains
  gene_expr_gain <- z_scored_expression %>%
    dplyr::filter(Hugo_Symbol == gene & expr_zscore > 2)
  
  expr_gain_mat <- table(is_plp = hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% z_scored_expression$Kids_First_Biospecimen_ID_normal] %in% plp_gene_bs,
                   is_alt = factor(hist_cohort$Kids_First_Biospecimen_ID_normal[hist_cohort$Kids_First_Biospecimen_ID_normal %in% z_scored_expression$Kids_First_Biospecimen_ID_normal] %in% gene_expr_gain$Kids_First_Biospecimen_ID_normal, levels = c("FALSE", "TRUE")))
  
  expr_gain_df[gene, c("n_no_plp_no_alt", "n_plp_no_alt", "n_no_plp_alt", "n_plp_alt")] = unlist(expr_gain_mat)
  expr_gain_df[gene, c("oddsRatio", "p", "CI_lower", "CI_upper")] <- unlist(fisher.test(expr_gain_mat)[c("estimate", "p.value", "conf.int")])


}  
```

Perform multiple test correction for each alteration type
```{r}
snv_df$adjusted_p <- p.adjust(snv_df$p, method = "BH")
cnv_loss_df$adjusted_p <- p.adjust(cnv_loss_df$p, method = "BH")
cnv_gain_df$adjusted_p <- p.adjust(cnv_gain_df$p, method = "BH")
loh_df$adjusted_p <- p.adjust(loh_df$p, method = "BH")
expr_loss_df$adjusted_p <- p.adjust(expr_loss_df$p, method = "BH")
expr_gain_df$adjusted_p <- p.adjust(expr_gain_df$p, method = "BH")

```

Create enrichment, p-value, logical significance, and label matrices for plotting
```{r}
alt_df <- snv_df %>%
  bind_rows(cnv_loss_df, cnv_gain_df, 
            loh_df, expr_loss_df, expr_gain_df) %>%
  dplyr::mutate(Hugo_Symbol = rep(genes, 6))

alterations = c("Oncogenic SNV", "CN loss", "CN gain",
                "LOH > 0.25", "TPM z-score < -2", "TPM z-score > 2")

enr_mat <- matrix(alt_df$oddsRatio, 
                  length(genes), length(alterations),
                  dimnames = list(genes, alterations))

p_mat <- matrix(alt_df$adjusted_p, 
                  length(genes), length(alterations),
                  dimnames = list(genes, alterations))

sig_mat <- ifelse(p_mat < 0.05, "*", "")

fill_mat <- matrix(glue::glue("{round(enr_mat, 1)} {sig_mat}"), 
                   length(genes), length(alterations))
```

Plot somatic alteration enrichment heatmap
```{r}
# Plot enrichment heatmap
col_fun = colorRamp2(c(0, 6), c("white", "orangered"))

pdf(file.path(plot_dir, "somatic-alteration-enr-heatmap.pdf"),
      width = 5, height = 6)

enr_heatmap <- Heatmap(enr_mat,
                       name = "Odds ratio",
                       cluster_rows = F,
                       cluster_columns = F,
                       rect_gp = gpar(col = "black", lwd = 2),
                       col = col_fun,
                       cell_fun = function(j, i, x, y, width, height, fill) {
                       grid.text(sprintf("%s", fill_mat[i, j]), x, y, gp = gpar(fontsize = 10))
               })

print(enr_heatmap)

dev.off()
```

Create data frame of significant enrichments to plot
```{r}
sig_alt_df <- alt_df %>%
  dplyr::filter(adjusted_p < 0.05) %>%
  dplyr::select(Hugo_Symbol, -n_no_plp_alt, -n_no_plp_no_alt, everything()) %>%
  dplyr::rename(n_alt = n_plp_alt,
                n_no_alt = n_plp_no_alt) %>%
  dplyr::mutate(group = "Germline P-LP")

# create reference data frame with enrichment = 1 to plot against enrichment scores in patients w/ P-LP variants
df_no_plp <- sig_alt_df %>%
  dplyr::select(Hugo_Symbol, alteration, 
                n_no_plp_alt, n_no_plp_no_alt) %>%
  dplyr::rename(n_alt = n_no_plp_alt,
                n_no_alt = n_no_plp_no_alt) %>%
  dplyr::mutate(oddsRatio = 1,
                p = NA, 
                CI_lower = NA, CI_upper = NA, 
                adjusted_p = NA,
                group = "No germline P-LP")

# merge data frames, and create additional columns for plotting
df_all <- sig_alt_df %>%
  bind_rows(df_no_plp) %>%
  dplyr::mutate(perc = n_alt/(n_alt + n_no_alt) * 100,
                fraction = glue::glue("{n_alt}/{n_alt + n_no_alt}"),
                p_text = case_when(
                  group == "Germline P-LP" ~ glue::glue("p={signif(adjusted_p, digits = 2)}"),
                  TRUE ~ NA_character_
                )) %>%
  dplyr::mutate(gene_alt = glue::glue("{Hugo_Symbol}: {alteration}"))

```

Plot enrichment of somatic alterations in patients with P-LP variants relative to those without
```{r}
# Create CPG Odds Ratio plot 
enr_plot <- df_all %>% 
  ggplot(aes(x = group, y = log10(oddsRatio), label = p_text)) +
  geom_point(size = 3, color = "#00A087FF",
             show.legend = FALSE) + 
  geom_errorbar(aes(ymin = log10(CI_lower), ymax = log10(CI_upper)), width = 0.2, 
                show.legend = FALSE, color = "#00A087FF") +
  labs(y = "log10-OR (95% CI)", x = NULL) + 
  geom_text(y = 1.6, x = 2, hjust = 0, size = 4, fontface = 2) +
  coord_flip() +
  geom_hline(yintercept = log10(1), linetype = "dashed") + 
  facet_wrap(~gene_alt, nrow = 7, scale = "fixed") +
  expand_limits(y=0) +
  theme_Publication() +
  theme(plot.margin = unit(c(2,0.5,1,0.25), "lines"))

# Create % patients with CPG PLP plot separated by source call, and include fractions as text 
perc_plot <- df_all %>%
  ggplot(aes(x = perc, y = group, label = fraction, fill = group)) +
  geom_bar(stat = "identity", color = "black",
           show.legend = FALSE) + 
  geom_text(x = 85, hjust = 0, size = 4, fontface = 2) +
  labs(x = "% patients", y = NULL, fill = NULL) + 
  guides(fill = guide_legend(nrow = 1)) +
  facet_wrap(~gene_alt, nrow = 7, scale = "fixed") +
  scale_y_discrete(labels = c("Germline P-LP" = NULL,
                   "No germline P-LP" = NULL)) +
  expand_limits(x=2) +
  coord_cartesian(clip = 'off') +
  theme_Publication() +
  theme(plot.margin = unit(c(2,6,1,1), "lines"),
        legend.position = c(0.5, 1.07))


pdf(file.path(plot_dir, "sig-somatic-alteration-enr.pdf"),
     width = 7, height = 8)

ggarrange(enr_plot, perc_plot,
          nrow = 1, widths = c(2,2))

dev.off()
```

Write `alt_df` to output
```{r}
alt_df %>%
  dplyr::relocate(Hugo_Symbol) %>%
  write_tsv(file.path(results_dir, "somatic-alteration-enrichment.tsv"))
```
