---
title: "Assess cancer predisposition gene protein abundance in PBTA germline cohort"
output: 
  html_notebook:
  toc: TRUE
toc_float: TRUE
author: Ryan Corbett
---

This script calculates whole-protein and phosphoprotein abundance z-scores in CPTAC and HOPE cohorts, and integrates this data with PBTA germline data to identify any putative associations between pathogenic germline variation and differential protein/phosphoprotein expression 

# Load libraries
```{r load libraries, set up directories}
suppressPackageStartupMessages({
  library(tidyverse)
})

# Set up directories
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "two-hits")
input_dir <- file.path(analysis_dir, "input")
results_dir <- file.path(analysis_dir, "results")

# If the directory does not exist, create it
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

# set seed to reproduce plots
set.seed(2023)

```

Set file paths

```{r set file paths}

opc_hist_file <- file.path(root_dir, "analyses",
                           "collapse-tumor-histologies",
                           "input", "histologies.tsv")

hist_file <- file.path(root_dir, "analyses",
                           "collapse-tumor-histologies", "results", 
                           "germline-primary-plus-tumor-histologies-plot-groups-clin-meta.tsv")

plp_file <- file.path(root_dir, "analyses", 
                      "variant-distribution", "input", 
                      "pbta_germline_plp_calls_autogvp_abridged_gnomad_popmax.tsv")

plp_sv_file <- file.path(root_dir, "analyses", 
                         "germline-sv", "input", 
                         "pbta_germline_svs.tsv")

cpg_file <- file.path(root_dir, "analyses", "oncokb-annotation",
                             "input", "cpg.txt")

cptac_proteo_file <- file.path(input_dir, 
                               "cptac-protein-imputed-prot-expression-abundance.tsv.gz")
hope_proteo_file <- file.path(input_dir,
                              "hope-protein-imputed-prot-expression-abundance.tsv.gz")
hope_phospho_file <- file.path(input_dir, 
                               "hope-protein-imputed-phospho-expression-abundance.tsv.gz")

```

Create proteo and phospoproteo specific histologies files

```{r}

opc_hist <- read_tsv(opc_hist_file)

opc_hist_proteo <- opc_hist %>%
  dplyr::filter(experimental_strategy == "Whole Cell Proteomics") %>%
  dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID,
                sample_id,
                experimental_strategy, sample_type, tumor_descriptor,
                sub_cohort) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_proteo = Kids_First_Biospecimen_ID,
                sample_id_tumor = sample_id) %>%
  dplyr::mutate(match_id = str_replace(Kids_First_Biospecimen_ID_proteo, "_WCP", ""))
  
opc_hist_phospho <- opc_hist %>%
  dplyr::filter(experimental_strategy == "Phospho-Proteomics") %>%
  dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID,
                experimental_strategy, sample_type, tumor_descriptor,
                sub_cohort) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_phospho = Kids_First_Biospecimen_ID) %>%
  dplyr::mutate(match_id = str_replace(Kids_First_Biospecimen_ID_phospho, "_PHOS", ""))

```

Load project-specific histologies file and append proteo and phosphoproteo BS IDs

```{r}
hist <- read_tsv(hist_file) %>%
  left_join(opc_hist_proteo[,c("Kids_First_Biospecimen_ID_proteo",
                           "sub_cohort", "sample_id_tumor", "match_id")],
            by = "sample_id_tumor") %>%
  left_join(opc_hist_phospho[,c("Kids_First_Biospecimen_ID_phospho",
                           "match_id")],
            by = "match_id") %>%
  dplyr::filter(!is.na(Kids_First_Biospecimen_ID_proteo) | !is.na(Kids_First_Biospecimen_ID_phospho))

```

Load P/LP file and append relevant hist columns

```{r}
plp <- read_tsv(plp_file)

cpgs <- read_lines(cpg_file)
                   
plp_cpg_proteo <- plp %>%
  dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist$Kids_First_Biospecimen_ID_normal & gene_symbol_vep %in% cpgs) %>%
  left_join(hist[,c("Kids_First_Biospecimen_ID_normal", "Kids_First_Participant_ID", "sample_id_tumor", "plot_group", "molecular_subtype")])

```

```{r}
plp_sv <- read_tsv(plp_sv_file)

plp_sv_cpg_proteo <- plp_sv %>%
  dplyr::filter(Kids_First_Biospecimen_ID_normal %in% hist$Kids_First_Biospecimen_ID_normal) %>%
  left_join(hist[,c("Kids_First_Biospecimen_ID_normal", "Kids_First_Participant_ID", "sample_id_tumor", "plot_group", "molecular_subtype")])

```

Load cptac proteomics file. In the event of multiple columns with the same sample ID, calculate abundance averages for sample. 

```{r}

cptac_proteo <- read_tsv(cptac_proteo_file)

cptac_proteo <-  cptac_proteo %>%
  dplyr::select(-NP_id)

# change column names to sample IDs
colnames(cptac_proteo)[2:ncol(cptac_proteo)] <- hist$sample_id_tumor[match(colnames(cptac_proteo)[2:ncol(cptac_proteo)], hist$Kids_First_Biospecimen_ID_proteo)]

# Filter for samples in germline hist
cptac_proteo <- cptac_proteo[,grepl("GeneSymbol", colnames(cptac_proteo)) | colnames(cptac_proteo) %in% hist$sample_id_tumor]


```

Load HOPE proteomics file, and resolve duplicates as was done with CPTAC data

```{r}

hope_proteo <- read_tsv(hope_proteo_file) %>%
  distinct(GeneSymbol, .keep_all = T) %>%
  dplyr::select(-NP_id)

# change column names to sample IDs
colnames(hope_proteo)[2:ncol(hope_proteo)] <- hist$sample_id_tumor[match(colnames(hope_proteo)[2:ncol(hope_proteo)], hist$Kids_First_Biospecimen_ID_proteo)]

# filter for sample IDs in germline hist
hope_proteo <- hope_proteo[,grepl("GeneSymbol", colnames(hope_proteo)) | colnames(hope_proteo) %in% hist$sample_id_tumor]

```

Merge hope and cptac proteomics data. In cases of duplicated samples, we will determine correlation between hope vs. cptac counts; if results are highly correlated, we will take average of two data sets.  

```{r}

proteo_mat <- hope_proteo %>%
  left_join(cptac_proteo, by = "GeneSymbol") %>%
  column_to_rownames("GeneSymbol")

# Identify duplicated columns
duplicated_columns <- intersect(names(hope_proteo[,-1]), names(cptac_proteo[,-1]))

# calcualte correlation between hope and cptac results for duplicated samples
cor_df <- tibble(sample = duplicated_columns,
                 correlation = rep(0, length(duplicated_columns)))

for (sample in duplicated_columns) {
  
  sample_hope <- glue::glue("{sample}.x")
  sample_cptac <- glue::glue("{sample}.y")
  
  cor_df$correlation[cor_df$sample == sample] <- cor.test(proteo_mat[,sample_hope],
                                                          proteo_mat[,sample_cptac])$estimate
  
}

cor_df

```

HOPE and CPTAC data are highly positively correlated, so we will take average of two runs for each sample

```{r}
# Calculate average of duplicated columns
for (col_name in duplicated_columns) {
  proteo_mat[[col_name]] <- rowMeans(proteo_mat[, grepl(col_name, names(proteo_mat))], na.rm = TRUE)
}

# Remove duplicated columns
proteo_mat <- proteo_mat %>%
  dplyr::select(-ends_with(c(".x", ".y")))

```

Calculate protein expression z-scores and save to output

```{r}
log_proteo <- log2(proteo_mat + 1)

# Scale the gene values -- scale() works on the columns, hence the transpose
z_scored_proteo <- scale(t(log_proteo),
                            center = TRUE,
                            scale = TRUE)
z_scored_proteo <- as.data.frame(z_scored_proteo)

# Convert to gathered data frame and append BS normal ID
z_scored_proteo <- z_scored_proteo %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("GeneSymbol") %>%
  gather(key = "sample_id_tumor", value = "expr_zscore_proteo", -GeneSymbol) %>%
  left_join(hist[,c("Kids_First_Biospecimen_ID_normal", "sample_id_tumor")]) %>%
  distinct(sample_id_tumor, GeneSymbol, .keep_all = T) %>%
  dplyr::filter(GeneSymbol %in% cpgs) %>%
  write_tsv(file.path(results_dir, "cptac-hope-cpg-proteomics-zscores.tsv"))

```

Load HOPE phosphoproteomics data 

```{r}

hope_phospho <- read_tsv(hope_phospho_file) %>%
  dplyr::mutate(Gene_Site = glue::glue("{GeneSymbol}:{Site}"))

phospho_mat <- hope_phospho %>%
  distinct(GeneSymbol, Site, .keep_all = T) %>%
  dplyr::select(-GeneSymbol, -Peptide_sequence, -NP_id, -Peptide_res_num, -Site) %>%
  column_to_rownames("Gene_Site")

# Convert column names to sample id; either extract from current column name or match to hist with BS ID
names(phospho_mat) <- ifelse(grepl("Prot", names(phospho_mat)), 
                             unlist(lapply(strsplit(names(phospho_mat), "_"), function(x) x[[2]])),
                             hist$sample_id_tumor[match(colnames(phospho_mat), hist$Kids_First_Biospecimen_ID_phospho)])
names(phospho_mat) <- str_replace(names(phospho_mat), "UP", "")

# Remove samples with NA column name (not in germline cohort)
phospho_mat <- phospho_mat[,!is.na(names(phospho_mat)) & names(phospho_mat) %in% hist$sample_id_tumor]
```

Calculate phosphoproteomics z-scores

```{r}

log_phospho <- log2(phospho_mat + 1)

# Scale the gene values -- scale() works on the columns, hence the transpose
z_scored_phospho <- scale(t(log_phospho),
                            center = TRUE,
                            scale = TRUE)
z_scored_phospho <- as.data.frame(z_scored_phospho)

# convert to gathered data frame and append normal BS ID
z_scored_phospho <- z_scored_phospho %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Gene_Site") %>%
  gather(key = "sample_id_tumor", value = "expr_zscore_phospho", -Gene_Site) %>%
  left_join(hope_phospho[,c("Gene_Site", "GeneSymbol")]) %>%
  left_join(hist[,c("Kids_First_Biospecimen_ID_normal", "sample_id_tumor")]) %>%
  dplyr::select(Kids_First_Biospecimen_ID_normal, sample_id_tumor,
                GeneSymbol, Gene_Site, expr_zscore_phospho) %>%
  distinct(sample_id_tumor, Gene_Site, .keep_all = T) %>%
  dplyr::filter(GeneSymbol %in% cpgs) %>%
  write_tsv(file.path(results_dir, "hope-cpg-phosphoproteomics-zscores.tsv"))

```

Append protein and phosphoprotein expression z-scores to plp file and save to output

```{r}

plp_cpg_proteo <- plp_cpg_proteo %>%
  dplyr::rename(GeneSymbol = gene_symbol_vep) %>%
  left_join(z_scored_proteo, by = c("Kids_First_Biospecimen_ID_normal", "sample_id_tumor", "GeneSymbol")) %>%
  left_join(z_scored_phospho, by = c("Kids_First_Biospecimen_ID_normal", "sample_id_tumor", "GeneSymbol")) %>%
  dplyr::filter(!is.na(expr_zscore_proteo) | !is.na(expr_zscore_phospho)) %>%
  distinct(Kids_First_Biospecimen_ID_normal, GeneSymbol, Gene_Site, .keep_all = T) %>%
  dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID_normal,
                chr, start, ref, alt, GeneSymbol,
                autogvp_call, autogvp_call_reason,
                sample_id_tumor, plot_group, molecular_subtype,
                expr_zscore_proteo, Gene_Site, expr_zscore_phospho)

write_tsv(plp_cpg_proteo,
          file.path(results_dir, "germline-somatic-proteomics-phosphoproteomics-cpgs.tsv"))

```


```{r}

plp_sv_cpg_proteo <- plp_sv_cpg_proteo %>%
  dplyr::rename(GeneSymbol = Hugo_Symbol_cpg) %>%
  left_join(z_scored_proteo, by = c("Kids_First_Biospecimen_ID_normal", "sample_id_tumor", "GeneSymbol")) %>%
  left_join(z_scored_phospho, by = c("Kids_First_Biospecimen_ID_normal", "sample_id_tumor", "GeneSymbol")) %>%
  dplyr::filter(!is.na(expr_zscore_proteo) | !is.na(expr_zscore_phospho)) %>%
  distinct(Kids_First_Biospecimen_ID_normal, GeneSymbol, Gene_Site, .keep_all = T) %>%
  dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID_normal,
                Chromosome, Start, End, Length, Type, GeneSymbol,
                Classification, Caller,
                sample_id_tumor, plot_group, molecular_subtype,
                expr_zscore_proteo, Gene_Site, expr_zscore_phospho)

write_tsv(plp_sv_cpg_proteo,
          file.path(results_dir, "germline-sv-somatic-proteomics-phosphoproteomics-cpgs.tsv"))

```